<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>omniplot.chipseq API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>omniplot.chipseq</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from typing import List,Dict,Optional,Union,Any,Iterable
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

import os 
import scipy.stats
from scipy.spatial.distance import squareform, pdist
from scipy.spatial import distance
from joblib import Parallel, delayed
import itertools as it
from joblib.externals.loky import get_reusable_executor
from omniplot.chipseq_utils import (_stitching_for_pyrange, 
                                    gff_parser, 
                                    _read_peaks, 
                                    _read_bw, 
                                    _calc_pearson,
                                    _stitching,
                                    _read_tss,
                                    _remove_close_to_tss,
                                    _find_extremes,
                                    _readgff,
                                    _readgff2,
                                    _readgtf,
                                    _readgtf2,
                                    _read_and_reshape_bw,
                                    _read_bw_stats,
                                    _readgff_transcripts,
                                    _read_transcripts)
sns.set_theme(font=&#34;Arial&#34;, style={&#39;grid.linestyle&#39;: &#34;&#34;,&#39;axes.facecolor&#39;: &#39;whitesmoke&#39;})
import random
import string
from matplotlib.colors import LogNorm
from sklearn.decomposition import PCA, NMF, LatentDirichletAllocation
from scipy.stats import fisher_exact
from scipy.stats import zscore
from sklearn.cluster import KMeans
from omniplot.utils import _optimal_kmeans, _save
import time
import pyranges as pr
import warnings
import sys

try:
    import pyBigWig as pwg
except ImportError as e:
    print(&#34;If you want to use chipseq functions. You need to install pyBigWig.&#34;)
except Exception as err:
    print(f&#34;Unexpected {err=}, {type(err)=}&#34;)
    raise

def plot_bigwig(files: dict, 
                bed: Union[str, list], 
                gff: str,
                step: int=100,
                stack_regions: str=&#34;horizontal&#34;, 
                highlight: Union[str, list]=[],
                highlightname=&#34;&#34;,
                n_jobs=-1,
                save: str=&#34;&#34;)-&gt;dict:
    &#34;&#34;&#34;
    Plotting bigwig files in specified genomic regions in the bed file.  
    
    Parameters
    ----------
    files : dict
        A dictionary whose keys are sample names and values are file names 
    bed : str
        A bed file name containing the list of genomic regions
    gff : str
        A gff3 file with a corresponding genome version
    step: int
        A bin size to reduce the bigwig signal resolution.
    show : bool
        Whether or not to show the figure.
    Returns
    -------
    {&#34;ax&#34;:axes,&#34;values&#34;:mat,&#34;genes&#34;:geneset,&#34;positions&#34;:pos} : dict
        A dictionary containing matplot axes, signal values, geneset, and positions.
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34; 
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
    if type(bed)==str:
        posrange=[]
        pos=[]
        with open(bed) as fin:
            
            for l in fin:
                l=l.split()
                chrom, s, e=l[0],l[1],l[2]
                posrange.append(pr.from_dict({&#34;Chromosome&#34;: [chrom], &#34;Start&#34;: [int(s.replace(&#34;,&#34;,&#34;&#34;))], &#34;End&#34;: [int(e.replace(&#34;,&#34;,&#34;&#34;))]}))
                pos.append([chrom,int(s.replace(&#34;,&#34;,&#34;&#34;)),int(e.replace(&#34;,&#34;,&#34;&#34;))])  
    else:
        pos=bed
        
    if type(highlight)==str:
        _highlight=[]
        with open(bed) as fin:
            for l in fin:
                l=l.split()
                chrom, s, e=l[0],l[1],l[2]
                _highlight.append([chrom,int(s.replace(&#34;,&#34;,&#34;&#34;)),int(e.replace(&#34;,&#34;,&#34;&#34;))])
        highlight=_highlight
    
    time_start=time.time()
    if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
        gff_dict=_readgff2(gff, &#34;gene&#34;, others=[&#34;gene_name&#34;, &#34;Strand&#34;])
    elif gff.endswith(&#34;gtf&#34;):
        gff_dict=_readgtf2(gff, &#34;transcript&#34;, others=[&#34;gene_name&#34;, &#34;Strand&#34;])
    else:
        raise Exception(&#34;The gff/gtf file name is required to have either gff, gff3, or gtf extension.&#34;)
    gff_ob=pr.from_dict(gff_dict)
    
    #
    # if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
    #     gff_ob=pr.read_gff3(gff)
    #     gff_ob=gff_ob[gff_ob.Feature==&#34;gene&#34;]
    #     gffformat=&#34;gff&#34;
    # elif gff.endswith(&#34;gtf&#34;):
    #     gff_ob=pr.read_gtf(gff)
    #     gff_ob=gff_ob[gff_ob.Feature==&#34;transcript&#34;]
    #     gffformat=&#34;gtf&#34;
    print(&#34;Reading the gene annotation file took&#34;, time.time()-time_start)
    # geneset=[]
    # for chrom, start, end in pos:
    #     geneset.append(gff_ob.get_genes(chrom, start, end))
    time_start=time.time()
    geneset=[]
    _grs=Parallel(n_jobs=n_jobs)(delayed(gff_ob.overlap)(gr) for gr in posrange)
    # for gr in posrange:
    #     gr2=gff_ob.intersect(gr)
    for gr2 in _grs:
        #gr2=gff_ob.intersect(gr)
        if len(gr2)==0:
            geneset.append([])
        else:
            #print(gr2)
            # if gffformat==&#34;gff&#34;:
            #     gr2=gr2[gr2.Feature==&#34;gene&#34;]
            # elif gffformat==&#34;gtf&#34;:
            #     gr2=gr2[gr2.Feature==&#34;transcript&#34;]
                
            tmp=[]
            for gene_name, start, end, ori in zip(gr2.gene_name, gr2.Start, gr2.End, gr2.Strand):
                tmp.append([gene_name, start, end, ori])
            geneset.append(tmp)
    #print(geneset)
    #geneset=Parallel(n_jobs=-1)(delayed(gff_ob.get_genes)(chrom, start, end) for chrom, start, end in pos)
    print(&#34;Obtaining genes took&#34;, time.time()-time_start)
            
    time_start=time.time()
    mat={}
    for samplename, f in files.items():
        bw=pwg.open(f)
        #h, t=os.path.split(f)
        #samplename=t.replace(prefix, &#34;&#34;).replace(suffix_set, &#34;&#34;)
        mat[samplename]=[]
        
        for chrom, s, e in pos:
            val=bw.values(chrom, s, e)
            
            mat[samplename].append(val)
    
    
    
    print(&#34;Reading bigwig files took&#34;, time.time()-time_start)
    
    if stack_regions==&#34;vertical&#34;:
        fig, axes=plt.subplots(nrows=len(pos)*2,ncols=len(files),figsize=[10,10],gridspec_kw={
                           &#39;height_ratios&#39;: [2,1]*len(pos),&#34;hspace&#34;:0},
                       constrained_layout = True)
        if len(axes.shape)==1:
            axes=axes.reshape([-1,1])
        sample_order=sorted(mat.keys())
        ticks_labels=[]
        for si, sample in enumerate(sample_order):
            
            vals=mat[sample]
            ymax=0
            for posi, (genes, val) in enumerate(zip(geneset, vals)):
                if len(val)==0:
                    continue
                ax=axes[posi*2,si]
                chrom, s, e=pos[posi]
                val=np.array(val)
                vallen=val.shape[0]
                remains=vallen%step
                _e=e-remains
                x=np.arange(s, _e,step)
                #print(s, e, _e, vallen, step)
                val=val[:vallen-remains].reshape(-1, step).mean(axis=1)
                _ymax=np.amax(val)
                if _ymax &gt;  ymax:
                    ymax =_ymax
                if len(highlight)!=0:
                    hchrom, hs, he=highlight[posi]
                    ax.fill_between([hs, he], [_ymax, _ymax], alpha=0.5, color=&#34;r&#34;)
                ax.fill_between(x, val,edgecolor=&#39;face&#39;)
                
                
                if si==0:
                    ax.set_title(sample+&#34; &#34;+chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e))
                    ticks_labels.append([[np.amin(x),np.amax(x)],ax.get_xticks(), ax.get_xticklabels()])
                ax.set_xticks([],labels=[])
            occupied=[[] for i in range(8)]
        
        

        
        for posi in range(len(pos)):
            interval=pos[posi][2]-pos[posi][1]
            axes[posi*2+1,si].plot(ticks_labels[posi][0],[0,0],alpha=0)
            for genename, gs, ge,ori in geneset[posi]:
                slot=0
                for oi in range(len(occupied)):
                    oc=False
                    if len(occupied[oi])==0:
                        occupied[oi].append([gs, ge])
                        oc=False
                        slot=oi
                    else:
                        for tmps, tmpe in occupied[oi]:
                            if tmps&lt;=gs&lt;= tmpe or tmps&lt;=ge&lt;=tmpe:
                                oc=True
                        if oc==False:
                            occupied[oi].append([gs, ge])
                            slot=oi
                    if oc==False:
                        break
                if gs &lt; pos[posi][1]:
                    gs=pos[posi][1]
                if ge &gt; pos[posi][2]:
                    ge=pos[posi][2]
                axes[posi*2+1,si].plot([gs, ge], [slot*1,slot*1], color=&#34;gray&#34;)
                if ori==&#34;+&#34;:
                    axes[posi*2+1,si].plot([ge-interval/32, ge], [slot*1+1,slot*1], color=&#34;gray&#34;)
                elif ori==&#34;-&#34;:
                    axes[posi*2+1,si].plot([gs, gs+interval/32], [slot*1,slot*1+1], color=&#34;gray&#34;)
                
                axes[posi*2+1,si].text((gs+ge)/2, slot*1, genename, ha=&#34;center&#34;)
            for oi, ol in enumerate(occupied):
                if len(ol)==0:
                    maxslot=oi
                    break
            axes[posi*2+1,si].set_ylim(-0.5,maxslot+0.5)
            #axes[posi*2+1,si].ticklabel_format(useOffset=False)
            axes[posi*2+1,si].set_yticks([],labels=[])
            axes[posi*2+1,si].set_xticks(ticks_labels[posi][1])
            axes[posi*2+1,si].set_ylabel(&#34;Genes&#34;)
        if save !=&#34;&#34;:
            if save.endswith(&#34;.pdf&#34;) or save.endswith(&#34;.png&#34;) or save.endswith(&#34;.svg&#34;):
                h, t=os.path.splitext(save)
                plt.savefig(h+&#34;_plot&#34;+t)
            else:
                plt.savefig(save+&#34;_plot.pdf&#34;)
    else:
        fig, axes=plt.subplots(nrows=len(files)+1,ncols=len(pos),gridspec_kw={
                           &#39;height_ratios&#39;: [2]*len(files)+[1]})
        
        if len(axes.shape)==1:
            axes=axes.reshape([-1,1])
        sample_order=sorted(mat.keys())
        for si, sample in enumerate(sample_order):
            vals=mat[sample]
            ymax=0
            #print(ymax)
            for posi, (genes, val) in enumerate(zip(geneset, vals)):
                if len(val)==0:
                    continue
                ax=axes[si,posi]
                chrom, s, e=pos[posi]
                val=np.array(val)
                vallen=val.shape[0]
                remains=vallen%step
                _e=e-remains
                x=np.arange(s, _e,step)
                
                #print(s, e, _e, vallen, step)
                val=val[:vallen-remains].reshape(-1, step).mean(axis=1)
                _ymax=np.amax(val)
                if _ymax &gt;  ymax:
                    ymax =_ymax

                ax.fill_between(x, val,edgecolor=&#39;face&#39;)

                if posi==0:
                    ax.set_ylabel(sample)
    
                ax.set_xticks([],labels=[])
                if si==0:
                    ax.set_title(chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e))
                ax.grid(False)
                    
            occupied=[[] for i in range(8)]
            for posi in range(len(pos)):
                axes[si,posi].set_ylim(0,ymax*1.01)
                if posi &gt;0:
                    axes[si,posi].set_yticks([],labels=[])
        
        for posi in range(len(pos)):
            interval=pos[posi][2]-pos[posi][1]
            for genename, gs, ge,ori in geneset[posi]:
                slot=0
                for oi in range(len(occupied)):
                    oc=False
                    if len(occupied[oi])==0:
                        occupied[oi].append([gs, ge])
                        oc=False
                        slot=oi
                    else:
                        for tmps, tmpe in occupied[oi]:
                            if tmps&lt;=gs&lt;= tmpe or tmps&lt;=ge&lt;=tmpe:
                                oc=True
                        if oc==False:
                            occupied[oi].append([gs, ge])
                            slot=oi
                    if oc==False:
                        break
                if gs &lt; pos[posi][1]:
                    gs=pos[posi][1]
                if ge &gt; pos[posi][2]:
                    ge=pos[posi][2]
                axes[si+1,posi].plot([gs, ge], [slot*1,slot*1], color=&#34;gray&#34;)
                if ori==&#34;+&#34;:
                    axes[si+1,posi].plot([ge-interval/32, ge], [slot*1+0.1,slot*1], color=&#34;gray&#34;)
                elif ori==&#34;-&#34;:
                    axes[si+1,posi].plot([gs, gs+interval/32], [slot*1,slot*1+0.1], color=&#34;gray&#34;)
                
                axes[si+1,posi].text((gs+ge)/2, slot*1, genename, ha=&#34;center&#34;)
                #axes[si+1,posi].grid(False)
            for oi, ol in enumerate(occupied):
                if len(ol)==0:
                    maxslot=oi
                    break
            axes[si+1,posi].set_ylim(-0.5,maxslot+0.5)
            axes[si+1,posi].ticklabel_format(useOffset=False)
            axes[si+1,posi].set_yticks([],labels=[])
            
        
        plt.tight_layout(h_pad=-1,w_pad=-1)
        _save(save, &#34;plot&#34;)
        
    return {&#34;ax&#34;:axes,&#34;values&#34;:mat,&#34;genes&#34;:geneset,&#34;positions&#34;:pos}


def plot_bigwig_correlation(files: dict, 
                            chrom: str=&#34;chr1&#34;,
                            step: int=1000,
                            palette: str=&#34;coolwarm&#34;,
                            figsize: list=[6,6],
                            show_val: bool=True,
                            clustermap_param: dict={},
                            peakfile: str=&#34;&#34;,
                            #show_val: bool=False,
                            show: bool=False,
                            n_jobs: int=-1, save: str=&#34;&#34;) -&gt; Dict:
    
    &#34;&#34;&#34;
    Calculate pearson correlations and draw a heatmap for bigwig files.  
    
    Parameters
    ----------
    files : dict
        A dictionary whose keys are sample names and values are bigwig file names 
    chrom : str, optional
        A chromsome to cal
    step: int, optional
        A bin size to reduce the bigwig signal resolution.
    palette: str, optional
    show_val: bool, optional
        Whether or not to show correlation values on the heatmap
    peakfile: str, optional
        A peak file/bed file containing genome regions of interest. If given this option, the correlation will be calculated using the genome regions listed in this file.
    show : bool, optional
        Whether or not to show the figure.
    n_jobs : int, optional
        The number of threads to use. It will use all available threads by default. 
    Returns
    -------
    {&#34;ax&#34;:axes,&#34;values&#34;:mat,&#34;genes&#34;:geneset,&#34;positions&#34;:pos} : dict
        A dictionary containing matplot axes, signal values, geneset, and positions.
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34;
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
    
    mat=[]
    samples=[]
    bw=pwg.open(list(files.values())[0])
    chrom_sizes=bw.chroms()
    
    
        
    
    if peakfile!=&#34;&#34;:
        # peaks=read_peaks(peakfile)
        # for k, f in files.items():
        #     bw=pwg.open(f)
        #     tmp=[]
        #     for chrom, se in peaks.items():
        #         for s, e in se:
        #             val=bw.stats(chrom,s,e, exact=True)[0]
        #             if val==None:
        #                 val=0
        #         tmp.append(val)
        #     mat.append(tmp)
        #     samples.append(k)
        peaks=_read_peaks(peakfile)
        for k, f in files.items():
            bw=pwg.open(f)
            
            tmp=Parallel(n_jobs=n_jobs)(delayed(_read_bw_stats)(chrom, s, e, f) for chrom, s, e in peaks)
            mat.append(tmp)
            samples.append(k)
        
    elif chrom !=&#34;all&#34;:
        chrom_size=chrom_sizes[chrom]
        
        samples=[]
        mat=[]
        #samples=list(files.keys())
        #mat=Parallel(n_jobs=n_jobs)(delayed(read_bw)(files[s],chrom, chrom_size, step) for s in samples)
        for k, f in files.items():
            bw=pwg.open(f)
            val=np.array(bw.values(chrom,0,chrom_size))
            rem=chrom_size%step
            val=val[:val.shape[0]-rem]
            print(val.shape)
            val=np.nan_to_num(val)
            val=val.reshape([-1,step]).mean(axis=1)
            print(val.shape)
            mat.append(val/np.sum(val))
            samples.append(k)
    elif chrom==&#34;all&#34;:
        chroms=chrom_sizes.keys()
        for k, f in files.items():
            
            #tmp=Parallel(n_jobs=n_jobs)(delayed(read_bw)(f, chrom, 0, chrom_sizes[chrom], f) for chrom, s, e in chroms)
            bw=pwg.open(f)
            tmp=[]
            for chrom in chroms:
                chrom_size=chrom_sizes[chrom]
                val=np.array(bw.values(chrom,0,chrom_size))
                rem=chrom_size%step
                val=val[:val.shape[0]-rem]
                val=np.nan_to_num(val)
                val=val.reshape([-1,step]).mean(axis=1)
                tmp.append(val/np.sum(val))
            
            samples.append(k)
            mat.append(np.concatenate(tmp))
    else:
        raise Exception(&#34;please specify chrom or peakfile options.&#34;)
    
    mat=np.array(mat)
    dmat=Parallel(n_jobs=-1)(delayed(_calc_pearson)(ind, mat) for ind in list(it.combinations(range(mat.shape[0]), 2)))
    dmat=np.array(dmat)
    get_reusable_executor().shutdown(wait=True)
    dmat=squareform(dmat)
    print(dmat)
    dmat+=np.identity(dmat.shape[0])
    g=sns.clustermap(data=dmat,xticklabels=samples,yticklabels=samples,
               method=&#34;ward&#34;, cmap=palette,
               col_cluster=True,
               row_cluster=True,
               figsize=figsize,
               rasterized=True,
               #cbar_kws={&#34;label&#34;:&#34;Pearson correlation&#34;}, 
               annot=show_val,
               **clustermap_param)
    g.cax.set_ylabel(&#34;Pearson correlation&#34;, rotation=-90,va=&#34;bottom&#34;)
    plt.setp(g.ax_heatmap.get_yticklabels(), rotation=0)  # For y axis
    plt.setp(g.ax_heatmap.get_xticklabels(), rotation=90) # For x axis
    
    
    _save(save, &#34;correlation&#34;)
    if show:
        plt.show()
    return {&#34;correlation&#34;: dmat,&#34;samples&#34;:samples}
    
def call_superenhancer(bigwig: str, 
                       bed: str,
                       stitch=25000,
                       tss_dist: int=0,
                       plot_signals=False,
                       gff: str=&#34;&#34;,
                       closest_genes: bool=False,
                       nearest_k: int=1,
                       go_analysis: bool=False,
                       gonames: list=[&#34;GO_Biological_Process_2021&#34;,&#34;Reactome_2022&#34;,&#34;WikiPathways_2019_Human&#34;],
                       sample: str=&#34;Sample&#34;,
                       n_jobs: int=12,
                       save: str=&#34;&#34;)-&gt;Dict:
    &#34;&#34;&#34;
    Find super enhancers and plot an enhancer rank.  
    
    Parameters
    ----------
    bigwig : str
        A bigwig file.
    peakfile : str
        A peak file.
    tss_dist: int, optional
        A distance from transcriptional start sites. If this option is more than 0, you need provide gff option too.
    stitch: int, optional
        The maximum distance of enhancers to stitch together.
    plot_signals : bool, optional
        Drawing super enhancer regions +- 10kb
    gff: str, optional (required when tss_dist is provided)
        A gff or gtf file that contains gene annotations. You need to chose a correct genome version of a gff/gtf file 
    
    closest_genes: bool=False,
        If this is True, closest genes of SE will be shown, instead of genomic regions. You can specify the number 
        of closest genes by &#34;nearest_k&#34; option.
    nearest_k: int, optional
        The number of closest genes to be included. Default: 1.
        
    go_analysis: bool, optional
        GSEA analysis based on the rank of SEs.
        
    gonames: list=[&#34;GO_Biological_Process_2021&#34;,&#34;Reactome_2022&#34;,&#34;WikiPathways_2019_Human&#34;]
        Gene sets for GSEA analysis.
    
    sample: str,optional
        The sample name to appear in the signal plot
    
    Returns
    -------
        dict {&#34;ax&#34;: ax, &#34;positions&#34;:pos, &#34;signals&#34;:y,&#34;nearest_genes&#34;:_nearestgenes}
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34; 
    
    bw=pwg.open(bigwig)
    chrom_sizes=bw.chroms()
    peaks=_read_peaks(bed)
    #stitched=stitching(peaks, stitch)
    
    
    if tss_dist != 0:
        start_time=time.time()
        stitched=pr.from_dict(_stitching_for_pyrange(peaks, stitch))
        
        if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
            gff_dict=_readgff(gff, &#34;transcript&#34;,tss_dist)
        elif gff.endswith(&#34;gtf&#34;):
            gff_dict=_readgtf(gff, &#34;transcript&#34;,tss_dist)
        else:
            raise Exception(&#34;The gff/gtf file name is required to have either gff, gff3, or gtf extension.&#34;)
        gffr=pr.from_dict(gff_dict)
        
        _stitched=stitched.subtract(gffr, nb_cpu=n_jobs)
        stitched={}
        for chrom, s, e in zip(_stitched.Chromosome, _stitched.Start, _stitched.End):
            if not chrom in stitched:
                stitched[chrom]=[]
            stitched[chrom].append([s,e])
        print(&#34;removing enhancers around TSS took &#34;, time.time()-start_time)
        # tss_pos=read_tss(tss,tss_dist)
        # stitched=remove_close_to_tss(stitched, tss_pos)
    else:
        stitched=_stitching(peaks, stitch)
        
        
    &#34;&#34;&#34;Obtaining closest genes&#34;&#34;&#34;
    _nearestgenes=[]
    if closest_genes==True:
        start_time=time.time()
        if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
            gff_dict=_readgff2(gff, &#34;gene&#34;, others=[&#34;gene_name&#34;])
        elif gff.endswith(&#34;gtf&#34;):
            gff_dict=_readgtf2(gff, &#34;transcript&#34;, others=[&#34;gene_name&#34;])
        else:
            raise Exception(&#34;The gff/gtf file name is required to have either gff, gff3, or gtf extension.&#34;)
        gffr=pr.from_dict(gff_dict)
        # grlist=[]
        # for chrom_se in pos:
        #     chrom, se=chrom_se.split(&#34;:&#34;)
        #     s, e=se.split(&#34;-&#34;)
        #     grlist.append(pr.from_dict({&#34;Chromosome&#34;:[chrom],&#34;Start&#34;:[int(s)],&#34;End&#34;:[int(e)]}))
        # if nearest_k==1:
        #     _nearestgenes=Parallel(n_jobs=-1)(delayed(gr.nearest)(gffr) for gr in grlist)
        # else:
        #     _nearestgenes=Parallel(n_jobs=-1)(delayed(gr.k_nearest)(gffr, k=nearest_k) for gr in grlist)
        grlist={&#34;Chromosome&#34;:[],&#34;Start&#34;:[],&#34;End&#34;:[]}
        stiched_dit={}
        for chrom, se in stitched.items():
            for s, e in se:
                stiched_dit[chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e)]=[]
                #grlist.append(pr.from_dict({&#34;Chromosome&#34;:[chrom],&#34;Start&#34;:[int(s)],&#34;End&#34;:[int(e)]}))
                grlist[&#34;Chromosome&#34;].append(chrom)
                grlist[&#34;Start&#34;].append(int(s))
                grlist[&#34;End&#34;].append(int(e))
                
        gr=pr.from_dict(grlist)
        _nearestgenes_gr=gr.k_nearest(gffr, k=nearest_k, nb_cpu=n_jobs)
        # _nearestgenes=[]
        # tmp=[]
        for i, (chrom, s, e, gene_name) in enumerate(zip(_nearestgenes_gr.Chromosome, _nearestgenes_gr.Start,_nearestgenes_gr.End,_nearestgenes_gr.gene_name)):
            stiched_dit[chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e)].append(gene_name)
        #
        #     tmp.append(gene_name)
        #     if i%nearest_k==0 and i&gt;0:
        #         _nearestgenes.append(tmp)
        #         tmp=[]
        # _nearestgenes.append(tmp)
        print(&#34;Finding the nearest genes took &#34;, time.time()-start_time)
        
    start_time=time.time()
    mat=[]
    pos=[]
    for chrom, se in stitched.items():
        for s, e in se:
            #print(chrom,s,e)
            val=bw.stats(chrom,s,e, type=&#34;sum&#34;, exact=True)[0]
            if val==None:
                val=0
            mat.append(val)
            pos.append(chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e))
            
    x, y, pos, sindex, srt=_find_extremes(mat, pos)
    b=-x[sindex]*np.amax(y) +y[sindex]
    print(&#34;Finding super enhancer took &#34;, time.time()-start_time, &#34;sec&#34;)
    
    # print(len(_nearestgenes), _nearestgenes[:10])
    # print(srt.shape, srt[:10])
    _nearestgenes=[stiched_dit[chromse] for chromse in pos]
    
        
    fig, ax=plt.subplots(figsize=[6,4])
    ax.scatter(x[:sindex], y[:sindex],s=5,rasterized=True)
    ax.scatter(x[sindex:], y[sindex:], color=&#34;r&#34;,s=5,rasterized=True)
    ax.plot(x,x*np.amax(y)+b, color=&#34;gray&#34;)
    
    if closest_genes==True:
        for j in range(5):
            ax.text(x[-j-1],y[-j-1], &#34;,&#34;.join(_nearestgenes[-j-1]))
    else:
        for j in range(5):
            ax.text(x[-j-1],y[-j-1], pos[-j-1])
    ax.text(0,y[-1]*0.7, &#34;SE No.: &#34;+str( y[sindex:].shape[0])+&#34;\nCutoff: &#34;+str(np.round(y[sindex],3)))
    ax.set_xlabel(&#34;Enhancers&#34;)
    ax.set_ylabel(&#34;Signal sum&#34;)
    ax.set_ylim(0,max(y)*1.05)
    
    tick_=x.shape[0]//3
    if 100&lt;tick_&lt;=1000:
        tick_=500
    elif 1000&lt;tick_&lt;=5000:
        tick_=1000
    elif 5000&lt;tick_&lt;=10000:
        tick_=5000
    else:
        tick_=10000
    ax.set_xticks(np.arange(0,x.shape[0], tick_)/x.shape[0],labels=np.arange(0,x.shape[0], tick_))
    plt.ticklabel_format(axis=&#34;y&#34;, style=&#34;sci&#34;, scilimits=(0,0))
    
    _save(save, &#34;rank&#34;)
    
    plt.subplots_adjust(right=0.620, bottom=0.130)
    
    
    stitched_out=os.path.splitext(bed)[0]+&#34;_SE.bed&#34;
    with open(stitched_out, &#34;w&#34;) as fout:
        
        if closest_genes==True:
            rank=1
            for _pos, _sig, _gr in zip(reversed(pos[sindex:]), reversed(y[sindex:]),reversed(_nearestgenes[sindex:])):
                chrom, se=_pos.split(&#34;:&#34;)
                s, e=se.split(&#34;-&#34;)
                fout.write(&#34;\t&#34;.join([chrom,str(s),str(e), &#34;,&#34;.join(_gr), str(_sig),str(rank)+&#34;\n&#34;]))
                rank+=1
        
        else:
            rank=1
            for _pos, _sig in zip(reversed(pos[sindex:]), reversed(y[sindex:])):
                chrom, se=_pos.split(&#34;:&#34;)
                s, e=se.split(&#34;-&#34;)
                fout.write(&#34;\t&#34;.join([chrom,str(s),str(e), str(_sig),str(rank)+&#34;\n&#34;]))
                rank+=1
    print(&#34;The SE file was written in &#34;+stitched_out)
    if go_analysis==True:
        print(&#34;Now, doing GO anlyses..&#34;)
        start_time=time.time()
        try:
            import gseapy
        except ImportError:
            raise ImportError(&#39;If you want GO analysis, you need to install gseapy. Try &#34;pip install gseapy&#34;.&#39;)
        godf={0:[],1:[]}
        y=(y-np.mean(y))/np.std(y)
        for _sig, genes in zip(reversed(y), reversed(_nearestgenes)):
            for gene in list(genes):
                godf[0].append(gene)
                godf[1].append(_sig)
        #print(godf)
        godf=pd.DataFrame(godf)
        godf=godf.set_index(0)
        for name in gonames:
            gs_res = gseapy.prerank(rnk=godf, # or data=&#39;./P53_resampling_data.txt&#39;
                         gene_sets=name,
                             threads=4,
                             min_size=5,
                             max_size=1000,
                             permutation_num=100, # reduce number to speed up testing
                             outdir=None, # don&#39;t write to disk
                             seed=6,
                             verbose=True,) # see what&#39;s going on behind the scenes)
            #print(gs_res.res2d)
            for index, row in gs_res.res2d.iterrows():
                #print(row)
                if row[&#34;FDR q-val&#34;] &lt;0.05:
                    t=row.Term
                    gseapy.gseaplot(rank_metric=gs_res.ranking,
                             term=t,
                             cmap=plt.cm.seismic,
                             **gs_res.results[t],
                             ofname=os.path.splitext(bed)[0]+&#34;_&#34;+name+&#34;_gsea.pdf&#34;
                             )
        print(&#34;GO analysis took &#34;, time.time()-start_time)
    if plot_signals:
        
        if gff==&#34;&#34;:
            raise Exception(&#34;Please provide a gff file to draw signals.&#34;)
        letters = string.ascii_lowercase
        tmpfile=&#39;&#39;.join(random.choice(letters) for i in range(10))+&#34;.bed&#34;
        highlights=[]
        with open(tmpfile, &#34;w&#34;) as fout:
            rank=1
            for pos, sig in zip(reversed(pos[sindex:]), reversed(y[sindex:])):
                chrom, se=pos.split(&#34;:&#34;)
                s, e=se.split(&#34;-&#34;)
                fout.write(&#34;\t&#34;.join([chrom,str(int(s)-10000),str(int(e)+10000), str(sig),str(rank)+&#34;\n&#34;]))
                highlights.append([chrom,int(s),int(e)])
                rank+=1
                if rank==6:
                    break
                    
        plot_bigwig(files={sample: bigwig}, 
                    bed=tmpfile, gff=gff, 
                    step=100,
                    stack_regions=&#34;vertical&#34;,
                    highlight=highlights,n_jobs=n_jobs, save=save)
        os.remove(tmpfile)
    return {&#34;ax&#34;: ax, &#34;positions&#34;:pos, &#34;signals&#34;:y,&#34;nearest_genes&#34;:_nearestgenes}


def plot_average(files: dict, 
                 bed: Union[str,list], 
                 order: list=[], extend: int=500, 
                 palette: str=&#34;coolwarm&#34;,
                 binsize: int=10,
                 clustering: str=&#34;&#34;,
                 n_clusters: int=5,
                 minval: Optional[float]=None,
                 orientaion=False,
                 save=&#34;&#34;,
                 n_jobs: int=-1):
    &#34;&#34;&#34;
    Plotting bigwig files centered at peak regions.  
    
    Parameters
    ----------
    files : dict
        A dictionary whose keys are sample names and values are file names 
    bed : str
        A bed file name containing the list of genomic regions
    order : list
        A sample order to plot. if a clustering option (except kmeans_all) is chosen, the first sample is the target of clustering analysis. 
    binsize: int
        A bin size to reduce the bigwig signal resolution.
    clustering : str
        Availables: [&#34;kmeans&#34;, &#34;kmeans_all&#34;, &#34;kmeans_auto&#34;]
    n_clusters : int
        the number of clusters for clustering analysis. if you chose kmeans_auto, this option will be ignored.
    Returns
    -------
        dict {&#34;values&#34;:data,&#34;potisons&#34;:pos,&#34;labels&#34;:labels}
        values: dictionary containing sample names as the keys and signals as the values
        positions: genomic positions sorted by signals displayed in the plot
        labels: list of clustering labels if kmean clustering was performed
        
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34;
    if len(order)==0:
        order=list(files.keys())
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
            
    
    
    bw=pwg.open(files[order[0]])
    chrom_sizes=bw.chroms()
    
    if type(bed)==str:
        
        pos=[]
        bedchroms=set()
        with open(bed) as fin:
            for l in fin:
                if l.startswith(&#34;#&#34;):
                    continue
                l=l.split()
                chrom, s, e=l[0],l[1],l[2]
                
                if not chrom in chrom_sizes:
                    raise Exception(&#34;There may be mismatches in chromosome names between the bed file and bigwig files.\n\
                the chromosome names in the bed file are: {}\n\
                those in the bigwig files are: {}&#34;.format(chrom, chrom_sizes.keys())) 
                
                s, e=int(s.replace(&#34;,&#34;,&#34;&#34;)),int(e.replace(&#34;,&#34;,&#34;&#34;))
                center=(s+e)//2
                cs=center-extend
                ce=center+extend
                if cs &lt;0 or chrom_sizes[chrom] &lt; ce:
                    continue
                pos.append([chrom,cs,ce])
                bedchroms.add(chrom)
    else:
        pos=bed
    
    
    data={}
    data_mean=[]
    
    for i, sample in enumerate(order):
        bigwig=files[sample]
        data[sample]=[]
        vals, mean=zip(*Parallel(n_jobs=n_jobs, prefer=&#34;threads&#34;)(delayed(_read_and_reshape_bw)(chrom, s, e, bigwig, binsize) for chrom, s, e in pos))
        
        
        data[sample]=np.nan_to_num(vals)
        if i==0:
            data_mean=mean
        # for chrom, s, e in pos:
        #
        #     val=bw.values(chrom, s, e)
        #     #print(val)
        #     val=np.array(val)
        #     #print(val.shape)
        #     val=val.reshape([-1,binsize]).mean(axis=1)
        #     #print(val.shape)
        #     data[sample].append(val)
        #     if i==0:
        #         data_mean.append(bw.stats(chrom, s, e, exact=True)[0])
    labels=[]
    if clustering==&#34;kmeans_all&#34;:
        mat=[]
        for sample in order:
            #print(np.array(data[sample]).shape)
            mat.append(data[sample])
        mat=np.concatenate(mat, axis=1)
        #print(mat.shape)
        mat=zscore(mat, axis=0)
        pca=PCA(n_components=5, random_state=1)
        xpca=pca.fit_transform(mat)
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(xpca)
        labels=kmX.labels_
    elif clustering==&#34;kmeans&#34;:
        mat=np.array(data[order[0]])
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(mat)
        labels=kmX.labels_
    elif clustering==&#34;kmeans_auto&#34;:
        mat=np.array(data[order[0]])
        optimalclusternum=_optimal_kmeans(mat, [2, 10])
        n_clusters=np.amax(optimalclusternum)
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(mat)
        labels=kmX.labels_
        
    sortindex=np.argsort(data_mean)[::-1]
    pos=np.array(pos)[sortindex]

    plotindex=np.arange(0, len(pos), len(pos)//1000)
    fig, axes=plt.subplots(ncols=len(order), figsize=[2*len(order), 6])
    if len(order)==1:
        axes=[axes]
    for i, (sample, ax) in enumerate(zip(order, axes)):
        vals=data[sample]

        vals=np.array(vals)

        vals=vals[sortindex]
        if clustering!=&#34;&#34;:
            
            _labels=np.array(labels)[sortindex]
            sortindex2=np.argsort(_labels)
            vals=vals[sortindex2]
            _pos=pos[sortindex2]
            _labels=_labels[sortindex2]
            _labels=_labels[plotindex]
            ulabel, clabel=np.unique(_labels, return_counts=True)

        vals=vals[plotindex]
        if np.sum(vals)&lt;=10**-6:
            raise Exception(&#34;Signals may be empty, or you may be using a wrong bed file.&#34;)
        if np.amax(vals) &lt;= 1  and np.amin(vals)&gt;=0:
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=0.00001))
        elif np.amin(vals) &lt; 0:
            vals=np.where(vals &lt; 0, 0, vals)
            warnings.warn(&#34;The bigwig contains negative values. Is this coverage file?&#34;)
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=1))
        elif minval!=None:
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=minval))
        else:
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=1))
        
        
        
        cbaxes = fig.add_axes([0.8*(i+1)/len(order), 0.93, 0.03, 0.06]) 
        cb = plt.colorbar(im, cax = cbaxes)
        
            
        # print(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4))
        # print(np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xticks(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4),labels=np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xlabel(&#34;Peak range [bp]&#34;)
        ax.set_title(sample)
        
        if clustering!=&#34;&#34;:
            total=0
            isep=0
            for label, count in zip(ulabel, clabel):
                
                total+=count
                if isep==ulabel.shape[0]-1:
                    break
                ax.plot([0, vals.shape[1]-1], [total,total], color=&#34;black&#34;)
                isep+=1
        ax.set_yticks([])
        ax.grid(False)
    _save(save, &#34;heatmap&#34;)
    fig, axes2=plt.subplots(ncols=len(order), figsize=[4*len(order), 3])
    if len(order)==1:
        axes2=[axes2]
    maxval=0
    for i, (sample, ax) in enumerate(zip(order, axes2)):
        vals=data[sample]
        vals=np.array(vals)
        
        labels=np.array(labels)
        if clustering !=&#34;&#34;:
            for label, count in zip(ulabel, clabel):
                
                _vals=np.mean(vals[labels==label],axis=0)
                #print(_vals.shape)
                #print(vals.shape)
                ax.plot(np.arange(-extend, extend, 2*extend//(_vals.shape[0])),_vals, label=label)
                _maxval=np.max(_vals)
                if maxval &lt; _maxval:
                    maxval=_maxval
            plt.legend()
        else:
            vals=vals.mean(axis=0)
            ax.plot(np.arange(-extend, extend, 2*extend//(vals.shape[0])),vals)
            _maxval=np.max(vals)
            if maxval &lt; _maxval:
                maxval=_maxval 
        if i==0:
            ax.set_ylabel(&#34;Mean signal value&#34;)
        # print(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4))
        # print(np.arange(-extend, extend+extend//2, extend//2))
        #ax.set_xticks(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4),labels=np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xlabel(&#34;Peak range [bp]&#34;)
        ax.set_title(sample)
    
    for ax in axes2:
        ax.set_ylim(0, maxval*1.05)
    
    _save(save, &#34;average&#34;)
    
    return {&#34;values&#34;:data,&#34;potisons&#34;:_pos,&#34;labels&#34;:labels, &#34;axes&#34;:axes,&#34;axes2&#34;:axes2}



def _plot_genebody(files: dict,
                files_minus: Optional[dict]=None,
                 bed: Union[list, str]=&#34;&#34;,
                 gff: str=&#34;&#34;,
                 centered_at: str=&#34;TSS&#34;, 
                 order: list=[], 
                 extend: int=3000, 
                 palette: str=&#34;coolwarm&#34;,
                 binsize: int=10,
                 clustering: str=&#34;&#34;,
                 n_clusters: int=5,
                 save: str=&#34;&#34;,lognorm=True,n_jobs=-1)-&gt; Dict:
    
    &#34;&#34;&#34;
    Plotting bigwig files around gene bodies.  
    
    Parameters
    ----------
    files : dict
        A dictionary whose keys are sample names and values are file names 
    bed : str
        A bed file name containing the list of genomic regions
    order : list
        A sample order to plot. if a clustering option (except kmeans_all) is chosen, the first sample is the target of clustering analysis. 
    binsize: int
        A bin size to reduce the bigwig signal resolution.
    clustering : str
        Availables: [&#34;kmeans&#34;, &#34;kmeans_all&#34;, &#34;kmeans_auto&#34;]
    n_clusters : int
        the number of clusters for clustering analysis. if you chose kmeans_auto, this option will be ignored.
    Returns
    -------
        dict {&#34;values&#34;:data,&#34;potisons&#34;:pos,&#34;labels&#34;:labels, &#34;axes&#34;:axes, &#34;axes2&#34;:axes2}
        values: dictionary containing sample names as the keys and signals as the values
        positions: genomic positions sorted by signals displayed in the plot
        labels: list of clustering labels if kmean clustering was performed
        
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34;
    if len(order)==0:
        order=list(files.keys())
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
    bw=pwg.open(files[order[0]])
    chrom_sizes=bw.chroms()
    
    if len(bed)&gt;0 and len(gff)&gt;0:
        warnings.warn(&#34;Ignoring the gff file, since the bed file is provided.&#34;)
    
    if len(bed)&gt;0:
        oriset=set([&#34;+&#34;, &#34;-&#34;])
        bedchroms=set()
        if type(bed)==str and len(bed)&gt;0:
            with open(bed) as fin:
                for l in fin:
                    l=l.split()
                    if not l[-1] in oriset:
                        raise Exception(&#34;The bed file must contain orientation information at the last column. e.g., chr1\t10000\t20000\t.\t.\t+&#34;)
                    
                    if not l[0] in chrom_sizes:
                        raise Exception(&#34;There may be mismatches in chromosome names between the bed file and bigwig files.\n\
                    the chromosome names in the bed file are: {}\n\
                    those in the bigwig files are: {}&#34;.format(l[0], chrom_sizes.keys())) 
            posplus=[]
            posminus=[]
            seen=set()
            with open(bed) as fin:
                for l in fin:
                    l=l.split()
                    chrom, s, e, ori=l[0],l[1],l[2],l[-1]
                    tmp=&#34;_&#34;.join([chrom, s, e, ori])
                    if not tmp in seen:
                        seen.add(tmp)
                    else:
                        continue
                    
                    bedchroms.add(chrom)
                    if ori==&#34;+&#34;:
                        cs=int(s)-extend
                        ce=int(s)+extend
                        if cs &lt;0 or chrom_sizes[chrom] &lt; ce:
                            continue
                        posplus.append([chrom,cs,ce])
                    else:
                        cs=int(e)-extend
                        ce=int(e)+extend
                        if cs &lt;0 or chrom_sizes[chrom] &lt; ce:
                            continue
    
                        posminus.append([chrom,cs,ce])
        else:
            posplus=bed[0]
            posminus=bed[1]
    
    elif len(gff)&gt;0:
        transcripts=_readgff_transcripts(gff, extend)
    
    
    if len(order)==0:
        order=list(files.keys())
    data={}
    data_mean=[]
    start_time=time.time()
    
    for i, sample in enumerate(order):
        bigwig=files[sample]
        
        if len(gff)&gt;0:
            tid_list=[]
            pvals=[]
            tids=transcripts[&#34;+&#34;].keys()
            tid_list.extend(tids)
            pvals=Parallel(n_jobs=n_jobs)(delayed(_read_transcripts)(bigwig, transcripts[&#34;+&#34;][tid], binsize, extend)  for tid in tids)
            print(pvals[:10])
            if files_minus!=None:
                bigwig=files_minus[sample]
            mvals=[]
            tids=transcripts[&#34;-&#34;].keys()
            mvals=Parallel(n_jobs=n_jobs)(delayed(_read_transcripts)(bigwig, transcripts[&#34;-&#34;][tid], binsize, extend)  for tid in tids)
            tid_list.extend(tids)
            
        #bw=pwg.open(bigwig)
        else:
            pvals=[]
            pmean=[]
            data[sample]=[]
            pvals, pmean=zip(*Parallel(n_jobs=n_jobs)(delayed(_read_and_reshape_bw)(chrom, s, e, bigwig, binsize)  for chrom, s, e in posplus))
            
            if files_minus!=None:
                bigwig=files_minus[sample]
                #bw=pwg.open(bigwig)
                mvals, mmean=zip(*Parallel(n_jobs=n_jobs)(delayed(_read_and_reshape_bw)(chrom, s, e, bigwig, binsize)  for chrom, s, e in posminus))
            else:
                mvals, mmean=zip(*Parallel(n_jobs=n_jobs)(delayed(_read_and_reshape_bw)(chrom, s, e, bigwig, binsize)  for chrom, s, e in posminus))
        
        mvals=np.flip(mvals, axis=1)
        vals=np.nan_to_num(np.concatenate([pvals,mvals]))
        print(vals.shape)
        if i ==0:
            sums=np.sum(vals, axis=1)
            remove=sums &gt;np.quantile(sums, 0.1)
            print(np.sum(remove))
        vals=vals[remove]
        #vals=vals.reshape([-1,vals.shape[1]//binsize,binsize]).mean(axis=2)
        print(np.amax(sums),np.amin(sums),vals.shape)
        data[sample]=vals
        
        if i==0:
            data_mean=np.amax(vals,axis=1)
        # for chrom, s, e in posplus:
        #     val=bw.values(chrom, s, e)
        #     #print(val)
        #     val=np.array(val)
        #     #print(val.shape)
        #     val=val.reshape([-1,binsize]).mean(axis=1)
        #     #print(val.shape)
        #     data[sample].append(val)
        #     if i==0:
        #         mval=bw.stats(chrom, s, e, exact=True)[0]
        #         if mval==None:
        #             data_mean.append(0)
        #         else:
        #             data_mean.append(mval)
        #
        # for chrom, s, e in posminus:
        #     chromsize=chrom_sizes[chrom]
        #     val=bw.values(chrom, s, e)
        #     #print(val)
        #     val=np.array(val)[::-1]
        #     #print(val.shape)
        #     val=val.reshape([-1,binsize]).mean(axis=1)
        #     #print(val.shape)
        #     data[sample].append(val)
        #     if i==0:
        #         mval=bw.stats(chrom, s, e, exact=True)[0]
        #         if mval==None:
        #             data_mean.append(0)
        #         else:
        #             data_mean.append(mval)
    print(&#34;Reading bigwig file took &#34;, time.time()-start_time)
    labels=[]
    if clustering==&#34;kmeans_all&#34;:
        mat=[]
        for sample in order:
            print(np.array(data[sample]).shape)
            mat.append(data[sample])
        mat=np.concatenate(mat, axis=1)
        mat=np.nan_to_num(mat)
        print(mat.shape)
        mat=zscore(mat, axis=0)
        pca=PCA(n_components=5, random_state=1)
        xpca=pca.fit_transform(mat)
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(xpca)
        labels=kmX.labels_
    elif clustering==&#34;kmeans&#34;:
        mat=np.nan_to_num(np.array(data[order[0]]))
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(mat)
        labels=kmX.labels_
    elif clustering==&#34;kmeans_auto&#34;:
        mat=np.nan_to_num(np.array(data[order[0]]))
        optimalclusternum=_optimal_kmeans(mat, [2, 10])
        n_clusters=np.amax(optimalclusternum)
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(mat)
        labels=kmX.labels_
    if len(gff) &gt;0:
        pos=np.array(tid_list)
    else:
        pos=np.array(posplus+posminus)[remove]
    #data_mean=np.nan_to_num(data_mean)
    sortindex=np.argsort(data_mean)[::-1]
    print(sortindex)
    pos=pos[sortindex]
    print(pos[:20])
    plotindex=np.arange(0, len(pos), len(pos)//1000)
    fig, axes=plt.subplots(ncols=len(order), figsize=[2*len(order), 6])
    if len(order)==1:
        axes=[axes]
    for i, (sample, ax) in enumerate(zip(order, axes)):
        vals=data[sample]
        
        vals=np.array(vals)
        
        vals=vals[sortindex]
        vals=vals/np.amax(vals, axis=1)[:,None]
        ax.plot(vals[0])
        ax.plot(vals[1])
        ax.plot(vals[2])
        plt.show()
        sys.exit()
        

        if clustering!=&#34;&#34;:
            
            _labels=np.array(labels)[sortindex]
            sortindex2=np.argsort(_labels)
            vals=vals[sortindex2]
            _pos=pos[sortindex2]
            _labels=_labels[sortindex2]
            _labels=_labels[plotindex]
            ulabel, clabel=np.unique(_labels, return_counts=True)

        vals=vals[plotindex]
        print(vals.shape)
        if lognorm==True:
            
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=np.amax([1, np.quantile(sums,0.5)])))
        else:
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;)
        cbaxes = fig.add_axes([0.8*(i+1)/len(order), 0.93, 0.03, 0.06]) 
        cb = plt.colorbar(im, cax = cbaxes)
        
            
        # print(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4))
        # print(np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xticks(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4),labels=np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xlabel(&#34;Peak range [bp]&#34;)
        ax.set_title(sample)
        
        if clustering!=&#34;&#34;:
            total=0
            isep=0
            for label, count in zip(ulabel, clabel):
                
                total+=count
                if isep==ulabel.shape[0]-1:
                    break
                ax.plot([0, vals.shape[1]-1], [total,total], color=&#34;black&#34;)
                isep+=1
        ax.set_yticks([])
        ax.grid(False)
        
    _save(save, &#34;heatmap&#34;)

    fig, axes2=plt.subplots(ncols=len(order), figsize=[4*len(order), 3])
    if len(order)==1:
        axes2=[axes2]
    maxval=0
    for i, (sample, ax) in enumerate(zip(order, axes2)):
        vals=data[sample]
        vals=np.array(vals)
        labels=np.array(labels)
        if clustering !=&#34;&#34;:
            for label, count in zip(ulabel, clabel):
                
                _vals=np.mean(vals[labels==label],axis=0)
                print(_vals.shape)
                print(vals.shape)
                ax.plot(np.arange(-extend, extend, 2*extend//(_vals.shape[0])),_vals, label=label)
                _maxval=np.max(_vals)
                if maxval &lt; _maxval:
                    maxval=_maxval
            plt.legend()
        else:
            vals=vals.mean(axis=0)
            ax.plot(np.arange(-extend, extend, 2*extend//(vals.shape[0])),vals)
            _maxval=np.max(vals)
            if maxval &lt; _maxval:
                maxval=_maxval 
        if i==0:
            ax.set_ylabel(&#34;Mean signal value&#34;)
        # print(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4))
        # print(np.arange(-extend, extend+extend//2, extend//2))
        #ax.set_xticks(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4),labels=np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xlabel(&#34;Peak range [bp]&#34;)
        ax.set_title(sample)
    
    for ax in axes:
        ax.set_ylim(0, maxval*1.05)
    _save(save, &#34;average&#34;)
    return {&#34;values&#34;:data,&#34;potisons&#34;:pos,&#34;labels&#34;:labels, &#34;axes&#34;:axes, &#34;axes2&#34;:axes2}

def plot_bed_correlation(files:dict,
                         method: str=&#34;pearson&#34;,
                         palette: str=&#34;coolwarm&#34;,
                         figsize: list=[6,6], 
                         show_val: bool=True, 
                         clustermap_param: dict={},
                         step: int=100,
                         show: bool=False):
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
    bedintervals={}
    minmax={}
    for sample, bedfile in files.items():
        bedintervals[sample]={}
        with open(bedfile) as fin:
            for l in fin:
                if l.startswith(&#34;#&#34;):
                    continue
                l=l.split()
                chrom=l[0]
                s, e=int(l[1]), int(l[2])
                if not chrom in bedintervals[sample]:
                    bedintervals[sample][chrom]=[]
                if not chrom in minmax:
                    minmax[chrom]=[]
                bedintervals[sample][chrom].append([s,e])
                minmax[chrom].append(s)
                minmax[chrom].append(e)
    minmaxval={}
    for chrom, se in minmax.items():
        minmaxval[chrom]=[np.min(se), np.max(se)]
    
    mat=[]
    samples=[]
    for sample, chrom_intervals in bedintervals.items():
        print(sample)
        tmp=[]
        for chrom, intervals in chrom_intervals.items():
            tmp2=[0 for _ in range(minmaxval[chrom][0], minmaxval[chrom][1],step)]
            for s, e in intervals:
                s=s-minmaxval[chrom][0]
                e=e-minmaxval[chrom][0]
                for i in range(s, e,step):
                    tmp2[i//step]=1
            tmp.extend(tmp2)
        mat.append(tmp)
        samples.append(sample)
    mat=np.array(mat)
    mat = mat[:, np.any(mat, axis=0)]
    print(mat[:100,:100])
    if method==&#34;pearson&#34;:
        #dmat=Parallel(n_jobs=-1)(delayed(_calc_pearson)(ind, mat) for ind in list(it.combinations(range(mat.shape[0]), 2)))
        dmat=np.corrcoef(mat)
        dmat=np.array(dmat)
        #dmat=squareform(dmat)
        print(dmat)
        #dmat+=np.identity(dmat.shape[0])
    else:
        dmat=squareform(pdist(mat, method))
    
            
    
    g=sns.clustermap(data=dmat,xticklabels=samples,yticklabels=samples,
               method=&#34;ward&#34;, cmap=palette,
               col_cluster=True,
               row_cluster=True,
               figsize=figsize,
               rasterized=True,
               #cbar_kws={&#34;label&#34;:&#34;Pearson correlation&#34;}, 
               annot=show_val,
               **clustermap_param)
    if method==&#34;pearson&#34;:
        title=&#34;Pearson correlation&#34;
    else:
        title=method+&#34; distance&#34;
    g.cax.set_ylabel(title, rotation=-90,va=&#34;bottom&#34;)
    plt.setp(g.ax_heatmap.get_yticklabels(), rotation=0)  # For y axis
    plt.setp(g.ax_heatmap.get_xticklabels(), rotation=90) # For x axis
    if show:
        plt.show()
        
    return {&#34;correlation&#34;: dmat,&#34;samples&#34;:samples}
    
    
    
    
def frip_calc(files:dict,):
    
    pass
    
    
    
    

if __name__==&#34;__main__&#34;:
    #test=&#34;plot_bigwig&#34;
    test=&#34;plot_bigwig_correlation&#34;
    
    #test=&#34;plot_bigwig&#34;
    test=&#34;plot_bed_correlation&#34;
    
    test=&#34;plot_average&#34;
    #test=&#34;plot_genebody&#34;
    #test=&#34;call_superenhancer&#34;
    #test=&#34;plot_genebody&#34;
    import glob
    
    if test==&#34;plot_bed_correlation&#34;:
        fs= {&#34;KMT2A&#34;:&#34;/media/koh/grasnas/home/data/omniplot/KMT2A_ENCFF644EPI_srt.bed&#34;,
             &#34;KMT2B&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2B_ENCFF036WCY_srt.bed&#34;,
             &#34;HNF4G_rep1&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HNF4G_ENCFF413UQM_srt.bed&#34;,
             &#34;HNF4G_rep2&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HNF4G_ENCFF123FQW_srt.bed&#34;,
             &#34;HNF1A_rep1&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HNF1A_ENCFF696TGC_srt.bed&#34;,
             &#34;HNF1A_rep2&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HNF1A_ENCFF162SGM_srt.bed&#34;,}
        
        plot_bed_correlation(fs,step=5000, method=&#34;correlation&#34;)
        plt.show()
    elif test==&#34;plot_bigwig&#34;:
        fs= {&#34;KMT2A&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2A-human_ENCFF406SHU.bw&#34;,
             &#34;KMT2B&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2B-human_ENCFF709UTL.bw&#34;}
        bed=&#34;/media/koh/grasnas/home/data/omniplot/tmp3.bed&#34;
        gff=&#34;/media/koh/grasnas/home/data/omniplot/gencode.v40.annotation.gff3&#34;
        
        plot_bigwig(fs,bed,gff, step=10)
        plt.show()
    elif test==&#34;plot_bigwig_correlation&#34;:
        fs= {&#34;KMT2A&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2A-human_ENCFF406SHU.bw&#34;,
             &#34;KMT2B&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2B-human_ENCFF709UTL.bw&#34;,
             &#34;HNF1A_rep1&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_HNF1A-human_ENCFF397BTX.bw&#34;,
             &#34;HNF1A_rep2&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_HNF1A-human_ENCFF502ACF.bw&#34;,
             &#34;HNF4A_rep1&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_HNF4A-human_ENCFF467SMI.bw&#34;,
             &#34;HNF4A_rep2&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_HNF4A-human_ENCFF712VYA.bw&#34;,}
        
        plot_bigwig_correlation(fs,step=1000)
        plt.show()
    elif test==&#34;call_superenhancer&#34;:
        gff=&#34;/media/koh/grasnas/home/data/omniplot/gencode.v40.annotation.gff3&#34;
        
        gff=&#34;/media/koh/grasnas/home/data/omniplot/hg38.refGene.gtf&#34;
        gff=&#34;/media/koh/grasnas/home/data/omniplot/gencode.vM31.annotation.gff3&#34;
        f=&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2B-human_ENCFF709UTL.bw&#34;
        f=&#34;/media/koh/grasnas/home/data/omniplot/SRR620141_1_trimmed_srt_no_dups.bw&#34;
        peak=&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2B_ENCFF036WCY_srt.bed&#34;
        peak=&#34;/media/koh/grasnas/home/data/omniplot/peaks_mm39.bed&#34;
        #tss=&#34;/media/koh/grasnas/home/data/omniplot/gencode.v40.annotation_tss_srt.bed&#34;
        call_superenhancer(bigwig=f, bed=peak,tss_dist=5000,plot_signals=True , gff=gff,closest_genes=True,
                           go_analysis=False,nearest_k=3)
        plt.show()
    elif test==&#34;plot_average&#34;:
        
        peak=&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2B_ENCFF036WCY_srt.bed&#34;
        files={&#34;KMT2B&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2B-human_ENCFF709UTL.bw&#34;,
                            &#34;KMT2A&#34;:&#34;/media/koh/grasnas/home/data/omniplot/HepG2_KMT2A-human_ENCFF406SHU.bw&#34;}
        plot_average(files=files,
                            bed=peak,
                            order=[&#34;KMT2B&#34;,  &#34;KMT2A&#34;],clustering=&#34;kmeans&#34;
                            )
        plt.show()
    elif test==&#34;plot_genebody&#34;:

        gff=&#34;/media/koh/grasnas/home/data/omniplot/gencode.v40.annotation.gff3&#34;
        files={&#34;adrenal_grand&#34;:&#34;/media/koh/grasnas/home/data/omniplot/adrenal_grand_plus_ENCFF809PGE.bigWig&#34;}
        files_minus={&#34;adrenal_grand&#34;:&#34;/media/koh/grasnas/home/data/omniplot/adrenal_grand_minus_ENCFF896WDT.bigWig&#34;}
        _plot_genebody(files=files,files_minus=files_minus,
                            gff=gff,binsize=1,
                            order=[&#34;adrenal_grand&#34;]#,clustering=&#34;kmeans&#34;
                            )
        plt.show()
    #raise NotImplementedError(&#34;This function will plot ChIP-seq data.&#34;)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="omniplot.chipseq.call_superenhancer"><code class="name flex">
<span>def <span class="ident">call_superenhancer</span></span>(<span>bigwig: str, bed: str, stitch=25000, tss_dist: int = 0, plot_signals=False, gff: str = '', closest_genes: bool = False, nearest_k: int = 1, go_analysis: bool = False, gonames: list = ['GO_Biological_Process_2021', 'Reactome_2022', 'WikiPathways_2019_Human'], sample: str = 'Sample', n_jobs: int = 12, save: str = '') ‑> Dict[~KT, ~VT]</span>
</code></dt>
<dd>
<div class="desc"><p>Find super enhancers and plot an enhancer rank.
</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>bigwig</code></strong> :&ensp;<code>str</code></dt>
<dd>A bigwig file.</dd>
<dt><strong><code>peakfile</code></strong> :&ensp;<code>str</code></dt>
<dd>A peak file.</dd>
<dt><strong><code>tss_dist</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>A distance from transcriptional start sites. If this option is more than 0, you need provide gff option too.</dd>
<dt><strong><code>stitch</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>The maximum distance of enhancers to stitch together.</dd>
<dt><strong><code>plot_signals</code></strong> :&ensp;<code>bool</code>, optional</dt>
<dd>Drawing super enhancer regions +- 10kb</dd>
<dt><strong><code>gff</code></strong> :&ensp;<code>str</code>, optional <code>(required when tss_dist is provided)</code></dt>
<dd>A gff or gtf file that contains gene annotations. You need to chose a correct genome version of a gff/gtf file</dd>
<dt><strong><code>closest_genes</code></strong> :&ensp;<code>bool=False,</code></dt>
<dd>If this is True, closest genes of SE will be shown, instead of genomic regions. You can specify the number
of closest genes by "nearest_k" option.</dd>
<dt><strong><code>nearest_k</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>The number of closest genes to be included. Default: 1.</dd>
<dt><strong><code>go_analysis</code></strong> :&ensp;<code>bool</code>, optional</dt>
<dd>GSEA analysis based on the rank of SEs.</dd>
<dt><strong><code>gonames</code></strong> :&ensp;<code>list=["GO_Biological_Process_2021","Reactome_2022","WikiPathways_2019_Human"]</code></dt>
<dd>Gene sets for GSEA analysis.</dd>
<dt><strong><code>sample</code></strong> :&ensp;<code>str</code>,optional</dt>
<dd>The sample name to appear in the signal plot</dd>
</dl>
<h2 id="returns">Returns</h2>
<pre><code>dict {"ax": ax, "positions":pos, "signals":y,"nearest_genes":_nearestgenes}
</code></pre>
<h2 id="raises">Raises</h2>
<h2 id="notes">Notes</h2>
<h2 id="references">References</h2>
<h2 id="see-also">See Also</h2>
<h2 id="examples">Examples</h2></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def call_superenhancer(bigwig: str, 
                       bed: str,
                       stitch=25000,
                       tss_dist: int=0,
                       plot_signals=False,
                       gff: str=&#34;&#34;,
                       closest_genes: bool=False,
                       nearest_k: int=1,
                       go_analysis: bool=False,
                       gonames: list=[&#34;GO_Biological_Process_2021&#34;,&#34;Reactome_2022&#34;,&#34;WikiPathways_2019_Human&#34;],
                       sample: str=&#34;Sample&#34;,
                       n_jobs: int=12,
                       save: str=&#34;&#34;)-&gt;Dict:
    &#34;&#34;&#34;
    Find super enhancers and plot an enhancer rank.  
    
    Parameters
    ----------
    bigwig : str
        A bigwig file.
    peakfile : str
        A peak file.
    tss_dist: int, optional
        A distance from transcriptional start sites. If this option is more than 0, you need provide gff option too.
    stitch: int, optional
        The maximum distance of enhancers to stitch together.
    plot_signals : bool, optional
        Drawing super enhancer regions +- 10kb
    gff: str, optional (required when tss_dist is provided)
        A gff or gtf file that contains gene annotations. You need to chose a correct genome version of a gff/gtf file 
    
    closest_genes: bool=False,
        If this is True, closest genes of SE will be shown, instead of genomic regions. You can specify the number 
        of closest genes by &#34;nearest_k&#34; option.
    nearest_k: int, optional
        The number of closest genes to be included. Default: 1.
        
    go_analysis: bool, optional
        GSEA analysis based on the rank of SEs.
        
    gonames: list=[&#34;GO_Biological_Process_2021&#34;,&#34;Reactome_2022&#34;,&#34;WikiPathways_2019_Human&#34;]
        Gene sets for GSEA analysis.
    
    sample: str,optional
        The sample name to appear in the signal plot
    
    Returns
    -------
        dict {&#34;ax&#34;: ax, &#34;positions&#34;:pos, &#34;signals&#34;:y,&#34;nearest_genes&#34;:_nearestgenes}
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34; 
    
    bw=pwg.open(bigwig)
    chrom_sizes=bw.chroms()
    peaks=_read_peaks(bed)
    #stitched=stitching(peaks, stitch)
    
    
    if tss_dist != 0:
        start_time=time.time()
        stitched=pr.from_dict(_stitching_for_pyrange(peaks, stitch))
        
        if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
            gff_dict=_readgff(gff, &#34;transcript&#34;,tss_dist)
        elif gff.endswith(&#34;gtf&#34;):
            gff_dict=_readgtf(gff, &#34;transcript&#34;,tss_dist)
        else:
            raise Exception(&#34;The gff/gtf file name is required to have either gff, gff3, or gtf extension.&#34;)
        gffr=pr.from_dict(gff_dict)
        
        _stitched=stitched.subtract(gffr, nb_cpu=n_jobs)
        stitched={}
        for chrom, s, e in zip(_stitched.Chromosome, _stitched.Start, _stitched.End):
            if not chrom in stitched:
                stitched[chrom]=[]
            stitched[chrom].append([s,e])
        print(&#34;removing enhancers around TSS took &#34;, time.time()-start_time)
        # tss_pos=read_tss(tss,tss_dist)
        # stitched=remove_close_to_tss(stitched, tss_pos)
    else:
        stitched=_stitching(peaks, stitch)
        
        
    &#34;&#34;&#34;Obtaining closest genes&#34;&#34;&#34;
    _nearestgenes=[]
    if closest_genes==True:
        start_time=time.time()
        if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
            gff_dict=_readgff2(gff, &#34;gene&#34;, others=[&#34;gene_name&#34;])
        elif gff.endswith(&#34;gtf&#34;):
            gff_dict=_readgtf2(gff, &#34;transcript&#34;, others=[&#34;gene_name&#34;])
        else:
            raise Exception(&#34;The gff/gtf file name is required to have either gff, gff3, or gtf extension.&#34;)
        gffr=pr.from_dict(gff_dict)
        # grlist=[]
        # for chrom_se in pos:
        #     chrom, se=chrom_se.split(&#34;:&#34;)
        #     s, e=se.split(&#34;-&#34;)
        #     grlist.append(pr.from_dict({&#34;Chromosome&#34;:[chrom],&#34;Start&#34;:[int(s)],&#34;End&#34;:[int(e)]}))
        # if nearest_k==1:
        #     _nearestgenes=Parallel(n_jobs=-1)(delayed(gr.nearest)(gffr) for gr in grlist)
        # else:
        #     _nearestgenes=Parallel(n_jobs=-1)(delayed(gr.k_nearest)(gffr, k=nearest_k) for gr in grlist)
        grlist={&#34;Chromosome&#34;:[],&#34;Start&#34;:[],&#34;End&#34;:[]}
        stiched_dit={}
        for chrom, se in stitched.items():
            for s, e in se:
                stiched_dit[chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e)]=[]
                #grlist.append(pr.from_dict({&#34;Chromosome&#34;:[chrom],&#34;Start&#34;:[int(s)],&#34;End&#34;:[int(e)]}))
                grlist[&#34;Chromosome&#34;].append(chrom)
                grlist[&#34;Start&#34;].append(int(s))
                grlist[&#34;End&#34;].append(int(e))
                
        gr=pr.from_dict(grlist)
        _nearestgenes_gr=gr.k_nearest(gffr, k=nearest_k, nb_cpu=n_jobs)
        # _nearestgenes=[]
        # tmp=[]
        for i, (chrom, s, e, gene_name) in enumerate(zip(_nearestgenes_gr.Chromosome, _nearestgenes_gr.Start,_nearestgenes_gr.End,_nearestgenes_gr.gene_name)):
            stiched_dit[chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e)].append(gene_name)
        #
        #     tmp.append(gene_name)
        #     if i%nearest_k==0 and i&gt;0:
        #         _nearestgenes.append(tmp)
        #         tmp=[]
        # _nearestgenes.append(tmp)
        print(&#34;Finding the nearest genes took &#34;, time.time()-start_time)
        
    start_time=time.time()
    mat=[]
    pos=[]
    for chrom, se in stitched.items():
        for s, e in se:
            #print(chrom,s,e)
            val=bw.stats(chrom,s,e, type=&#34;sum&#34;, exact=True)[0]
            if val==None:
                val=0
            mat.append(val)
            pos.append(chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e))
            
    x, y, pos, sindex, srt=_find_extremes(mat, pos)
    b=-x[sindex]*np.amax(y) +y[sindex]
    print(&#34;Finding super enhancer took &#34;, time.time()-start_time, &#34;sec&#34;)
    
    # print(len(_nearestgenes), _nearestgenes[:10])
    # print(srt.shape, srt[:10])
    _nearestgenes=[stiched_dit[chromse] for chromse in pos]
    
        
    fig, ax=plt.subplots(figsize=[6,4])
    ax.scatter(x[:sindex], y[:sindex],s=5,rasterized=True)
    ax.scatter(x[sindex:], y[sindex:], color=&#34;r&#34;,s=5,rasterized=True)
    ax.plot(x,x*np.amax(y)+b, color=&#34;gray&#34;)
    
    if closest_genes==True:
        for j in range(5):
            ax.text(x[-j-1],y[-j-1], &#34;,&#34;.join(_nearestgenes[-j-1]))
    else:
        for j in range(5):
            ax.text(x[-j-1],y[-j-1], pos[-j-1])
    ax.text(0,y[-1]*0.7, &#34;SE No.: &#34;+str( y[sindex:].shape[0])+&#34;\nCutoff: &#34;+str(np.round(y[sindex],3)))
    ax.set_xlabel(&#34;Enhancers&#34;)
    ax.set_ylabel(&#34;Signal sum&#34;)
    ax.set_ylim(0,max(y)*1.05)
    
    tick_=x.shape[0]//3
    if 100&lt;tick_&lt;=1000:
        tick_=500
    elif 1000&lt;tick_&lt;=5000:
        tick_=1000
    elif 5000&lt;tick_&lt;=10000:
        tick_=5000
    else:
        tick_=10000
    ax.set_xticks(np.arange(0,x.shape[0], tick_)/x.shape[0],labels=np.arange(0,x.shape[0], tick_))
    plt.ticklabel_format(axis=&#34;y&#34;, style=&#34;sci&#34;, scilimits=(0,0))
    
    _save(save, &#34;rank&#34;)
    
    plt.subplots_adjust(right=0.620, bottom=0.130)
    
    
    stitched_out=os.path.splitext(bed)[0]+&#34;_SE.bed&#34;
    with open(stitched_out, &#34;w&#34;) as fout:
        
        if closest_genes==True:
            rank=1
            for _pos, _sig, _gr in zip(reversed(pos[sindex:]), reversed(y[sindex:]),reversed(_nearestgenes[sindex:])):
                chrom, se=_pos.split(&#34;:&#34;)
                s, e=se.split(&#34;-&#34;)
                fout.write(&#34;\t&#34;.join([chrom,str(s),str(e), &#34;,&#34;.join(_gr), str(_sig),str(rank)+&#34;\n&#34;]))
                rank+=1
        
        else:
            rank=1
            for _pos, _sig in zip(reversed(pos[sindex:]), reversed(y[sindex:])):
                chrom, se=_pos.split(&#34;:&#34;)
                s, e=se.split(&#34;-&#34;)
                fout.write(&#34;\t&#34;.join([chrom,str(s),str(e), str(_sig),str(rank)+&#34;\n&#34;]))
                rank+=1
    print(&#34;The SE file was written in &#34;+stitched_out)
    if go_analysis==True:
        print(&#34;Now, doing GO anlyses..&#34;)
        start_time=time.time()
        try:
            import gseapy
        except ImportError:
            raise ImportError(&#39;If you want GO analysis, you need to install gseapy. Try &#34;pip install gseapy&#34;.&#39;)
        godf={0:[],1:[]}
        y=(y-np.mean(y))/np.std(y)
        for _sig, genes in zip(reversed(y), reversed(_nearestgenes)):
            for gene in list(genes):
                godf[0].append(gene)
                godf[1].append(_sig)
        #print(godf)
        godf=pd.DataFrame(godf)
        godf=godf.set_index(0)
        for name in gonames:
            gs_res = gseapy.prerank(rnk=godf, # or data=&#39;./P53_resampling_data.txt&#39;
                         gene_sets=name,
                             threads=4,
                             min_size=5,
                             max_size=1000,
                             permutation_num=100, # reduce number to speed up testing
                             outdir=None, # don&#39;t write to disk
                             seed=6,
                             verbose=True,) # see what&#39;s going on behind the scenes)
            #print(gs_res.res2d)
            for index, row in gs_res.res2d.iterrows():
                #print(row)
                if row[&#34;FDR q-val&#34;] &lt;0.05:
                    t=row.Term
                    gseapy.gseaplot(rank_metric=gs_res.ranking,
                             term=t,
                             cmap=plt.cm.seismic,
                             **gs_res.results[t],
                             ofname=os.path.splitext(bed)[0]+&#34;_&#34;+name+&#34;_gsea.pdf&#34;
                             )
        print(&#34;GO analysis took &#34;, time.time()-start_time)
    if plot_signals:
        
        if gff==&#34;&#34;:
            raise Exception(&#34;Please provide a gff file to draw signals.&#34;)
        letters = string.ascii_lowercase
        tmpfile=&#39;&#39;.join(random.choice(letters) for i in range(10))+&#34;.bed&#34;
        highlights=[]
        with open(tmpfile, &#34;w&#34;) as fout:
            rank=1
            for pos, sig in zip(reversed(pos[sindex:]), reversed(y[sindex:])):
                chrom, se=pos.split(&#34;:&#34;)
                s, e=se.split(&#34;-&#34;)
                fout.write(&#34;\t&#34;.join([chrom,str(int(s)-10000),str(int(e)+10000), str(sig),str(rank)+&#34;\n&#34;]))
                highlights.append([chrom,int(s),int(e)])
                rank+=1
                if rank==6:
                    break
                    
        plot_bigwig(files={sample: bigwig}, 
                    bed=tmpfile, gff=gff, 
                    step=100,
                    stack_regions=&#34;vertical&#34;,
                    highlight=highlights,n_jobs=n_jobs, save=save)
        os.remove(tmpfile)
    return {&#34;ax&#34;: ax, &#34;positions&#34;:pos, &#34;signals&#34;:y,&#34;nearest_genes&#34;:_nearestgenes}</code></pre>
</details>
</dd>
<dt id="omniplot.chipseq.frip_calc"><code class="name flex">
<span>def <span class="ident">frip_calc</span></span>(<span>files: dict)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def frip_calc(files:dict,):
    
    pass</code></pre>
</details>
</dd>
<dt id="omniplot.chipseq.plot_average"><code class="name flex">
<span>def <span class="ident">plot_average</span></span>(<span>files: dict, bed: Union[str, list], order: list = [], extend: int = 500, palette: str = 'coolwarm', binsize: int = 10, clustering: str = '', n_clusters: int = 5, minval: Optional[float] = None, orientaion=False, save='', n_jobs: int = -1)</span>
</code></dt>
<dd>
<div class="desc"><p>Plotting bigwig files centered at peak regions.
</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>files</code></strong> :&ensp;<code>dict</code></dt>
<dd>A dictionary whose keys are sample names and values are file names</dd>
<dt><strong><code>bed</code></strong> :&ensp;<code>str</code></dt>
<dd>A bed file name containing the list of genomic regions</dd>
<dt><strong><code>order</code></strong> :&ensp;<code>list</code></dt>
<dd>A sample order to plot. if a clustering option (except kmeans_all) is chosen, the first sample is the target of clustering analysis.</dd>
<dt><strong><code>binsize</code></strong> :&ensp;<code>int</code></dt>
<dd>A bin size to reduce the bigwig signal resolution.</dd>
<dt><strong><code>clustering</code></strong> :&ensp;<code>str</code></dt>
<dd>Availables: ["kmeans", "kmeans_all", "kmeans_auto"]</dd>
<dt><strong><code>n_clusters</code></strong> :&ensp;<code>int</code></dt>
<dd>the number of clusters for clustering analysis. if you chose kmeans_auto, this option will be ignored.</dd>
</dl>
<h2 id="returns">Returns</h2>
<pre><code>dict {"values":data,"potisons":pos,"labels":labels}
values: dictionary containing sample names as the keys and signals as the values
positions: genomic positions sorted by signals displayed in the plot
labels: list of clustering labels if kmean clustering was performed
</code></pre>
<h2 id="raises">Raises</h2>
<h2 id="notes">Notes</h2>
<h2 id="references">References</h2>
<h2 id="see-also">See Also</h2>
<h2 id="examples">Examples</h2></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_average(files: dict, 
                 bed: Union[str,list], 
                 order: list=[], extend: int=500, 
                 palette: str=&#34;coolwarm&#34;,
                 binsize: int=10,
                 clustering: str=&#34;&#34;,
                 n_clusters: int=5,
                 minval: Optional[float]=None,
                 orientaion=False,
                 save=&#34;&#34;,
                 n_jobs: int=-1):
    &#34;&#34;&#34;
    Plotting bigwig files centered at peak regions.  
    
    Parameters
    ----------
    files : dict
        A dictionary whose keys are sample names and values are file names 
    bed : str
        A bed file name containing the list of genomic regions
    order : list
        A sample order to plot. if a clustering option (except kmeans_all) is chosen, the first sample is the target of clustering analysis. 
    binsize: int
        A bin size to reduce the bigwig signal resolution.
    clustering : str
        Availables: [&#34;kmeans&#34;, &#34;kmeans_all&#34;, &#34;kmeans_auto&#34;]
    n_clusters : int
        the number of clusters for clustering analysis. if you chose kmeans_auto, this option will be ignored.
    Returns
    -------
        dict {&#34;values&#34;:data,&#34;potisons&#34;:pos,&#34;labels&#34;:labels}
        values: dictionary containing sample names as the keys and signals as the values
        positions: genomic positions sorted by signals displayed in the plot
        labels: list of clustering labels if kmean clustering was performed
        
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34;
    if len(order)==0:
        order=list(files.keys())
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
            
    
    
    bw=pwg.open(files[order[0]])
    chrom_sizes=bw.chroms()
    
    if type(bed)==str:
        
        pos=[]
        bedchroms=set()
        with open(bed) as fin:
            for l in fin:
                if l.startswith(&#34;#&#34;):
                    continue
                l=l.split()
                chrom, s, e=l[0],l[1],l[2]
                
                if not chrom in chrom_sizes:
                    raise Exception(&#34;There may be mismatches in chromosome names between the bed file and bigwig files.\n\
                the chromosome names in the bed file are: {}\n\
                those in the bigwig files are: {}&#34;.format(chrom, chrom_sizes.keys())) 
                
                s, e=int(s.replace(&#34;,&#34;,&#34;&#34;)),int(e.replace(&#34;,&#34;,&#34;&#34;))
                center=(s+e)//2
                cs=center-extend
                ce=center+extend
                if cs &lt;0 or chrom_sizes[chrom] &lt; ce:
                    continue
                pos.append([chrom,cs,ce])
                bedchroms.add(chrom)
    else:
        pos=bed
    
    
    data={}
    data_mean=[]
    
    for i, sample in enumerate(order):
        bigwig=files[sample]
        data[sample]=[]
        vals, mean=zip(*Parallel(n_jobs=n_jobs, prefer=&#34;threads&#34;)(delayed(_read_and_reshape_bw)(chrom, s, e, bigwig, binsize) for chrom, s, e in pos))
        
        
        data[sample]=np.nan_to_num(vals)
        if i==0:
            data_mean=mean
        # for chrom, s, e in pos:
        #
        #     val=bw.values(chrom, s, e)
        #     #print(val)
        #     val=np.array(val)
        #     #print(val.shape)
        #     val=val.reshape([-1,binsize]).mean(axis=1)
        #     #print(val.shape)
        #     data[sample].append(val)
        #     if i==0:
        #         data_mean.append(bw.stats(chrom, s, e, exact=True)[0])
    labels=[]
    if clustering==&#34;kmeans_all&#34;:
        mat=[]
        for sample in order:
            #print(np.array(data[sample]).shape)
            mat.append(data[sample])
        mat=np.concatenate(mat, axis=1)
        #print(mat.shape)
        mat=zscore(mat, axis=0)
        pca=PCA(n_components=5, random_state=1)
        xpca=pca.fit_transform(mat)
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(xpca)
        labels=kmX.labels_
    elif clustering==&#34;kmeans&#34;:
        mat=np.array(data[order[0]])
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(mat)
        labels=kmX.labels_
    elif clustering==&#34;kmeans_auto&#34;:
        mat=np.array(data[order[0]])
        optimalclusternum=_optimal_kmeans(mat, [2, 10])
        n_clusters=np.amax(optimalclusternum)
        kmean = KMeans(n_clusters=n_clusters, random_state=0,n_init=10)
        kmX=kmean.fit(mat)
        labels=kmX.labels_
        
    sortindex=np.argsort(data_mean)[::-1]
    pos=np.array(pos)[sortindex]

    plotindex=np.arange(0, len(pos), len(pos)//1000)
    fig, axes=plt.subplots(ncols=len(order), figsize=[2*len(order), 6])
    if len(order)==1:
        axes=[axes]
    for i, (sample, ax) in enumerate(zip(order, axes)):
        vals=data[sample]

        vals=np.array(vals)

        vals=vals[sortindex]
        if clustering!=&#34;&#34;:
            
            _labels=np.array(labels)[sortindex]
            sortindex2=np.argsort(_labels)
            vals=vals[sortindex2]
            _pos=pos[sortindex2]
            _labels=_labels[sortindex2]
            _labels=_labels[plotindex]
            ulabel, clabel=np.unique(_labels, return_counts=True)

        vals=vals[plotindex]
        if np.sum(vals)&lt;=10**-6:
            raise Exception(&#34;Signals may be empty, or you may be using a wrong bed file.&#34;)
        if np.amax(vals) &lt;= 1  and np.amin(vals)&gt;=0:
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=0.00001))
        elif np.amin(vals) &lt; 0:
            vals=np.where(vals &lt; 0, 0, vals)
            warnings.warn(&#34;The bigwig contains negative values. Is this coverage file?&#34;)
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=1))
        elif minval!=None:
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=minval))
        else:
            im=ax.imshow(vals,aspect=&#34;auto&#34;, cmap=palette, interpolation=&#34;none&#34;,norm=LogNorm(vmin=1))
        
        
        
        cbaxes = fig.add_axes([0.8*(i+1)/len(order), 0.93, 0.03, 0.06]) 
        cb = plt.colorbar(im, cax = cbaxes)
        
            
        # print(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4))
        # print(np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xticks(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4),labels=np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xlabel(&#34;Peak range [bp]&#34;)
        ax.set_title(sample)
        
        if clustering!=&#34;&#34;:
            total=0
            isep=0
            for label, count in zip(ulabel, clabel):
                
                total+=count
                if isep==ulabel.shape[0]-1:
                    break
                ax.plot([0, vals.shape[1]-1], [total,total], color=&#34;black&#34;)
                isep+=1
        ax.set_yticks([])
        ax.grid(False)
    _save(save, &#34;heatmap&#34;)
    fig, axes2=plt.subplots(ncols=len(order), figsize=[4*len(order), 3])
    if len(order)==1:
        axes2=[axes2]
    maxval=0
    for i, (sample, ax) in enumerate(zip(order, axes2)):
        vals=data[sample]
        vals=np.array(vals)
        
        labels=np.array(labels)
        if clustering !=&#34;&#34;:
            for label, count in zip(ulabel, clabel):
                
                _vals=np.mean(vals[labels==label],axis=0)
                #print(_vals.shape)
                #print(vals.shape)
                ax.plot(np.arange(-extend, extend, 2*extend//(_vals.shape[0])),_vals, label=label)
                _maxval=np.max(_vals)
                if maxval &lt; _maxval:
                    maxval=_maxval
            plt.legend()
        else:
            vals=vals.mean(axis=0)
            ax.plot(np.arange(-extend, extend, 2*extend//(vals.shape[0])),vals)
            _maxval=np.max(vals)
            if maxval &lt; _maxval:
                maxval=_maxval 
        if i==0:
            ax.set_ylabel(&#34;Mean signal value&#34;)
        # print(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4))
        # print(np.arange(-extend, extend+extend//2, extend//2))
        #ax.set_xticks(np.arange(0, vals.shape[1]+vals.shape[1]//4, vals.shape[1]//4),labels=np.arange(-extend, extend+extend//2, extend//2))
        ax.set_xlabel(&#34;Peak range [bp]&#34;)
        ax.set_title(sample)
    
    for ax in axes2:
        ax.set_ylim(0, maxval*1.05)
    
    _save(save, &#34;average&#34;)
    
    return {&#34;values&#34;:data,&#34;potisons&#34;:_pos,&#34;labels&#34;:labels, &#34;axes&#34;:axes,&#34;axes2&#34;:axes2}</code></pre>
</details>
</dd>
<dt id="omniplot.chipseq.plot_bed_correlation"><code class="name flex">
<span>def <span class="ident">plot_bed_correlation</span></span>(<span>files: dict, method: str = 'pearson', palette: str = 'coolwarm', figsize: list = [6, 6], show_val: bool = True, clustermap_param: dict = {}, step: int = 100, show: bool = False)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_bed_correlation(files:dict,
                         method: str=&#34;pearson&#34;,
                         palette: str=&#34;coolwarm&#34;,
                         figsize: list=[6,6], 
                         show_val: bool=True, 
                         clustermap_param: dict={},
                         step: int=100,
                         show: bool=False):
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
    bedintervals={}
    minmax={}
    for sample, bedfile in files.items():
        bedintervals[sample]={}
        with open(bedfile) as fin:
            for l in fin:
                if l.startswith(&#34;#&#34;):
                    continue
                l=l.split()
                chrom=l[0]
                s, e=int(l[1]), int(l[2])
                if not chrom in bedintervals[sample]:
                    bedintervals[sample][chrom]=[]
                if not chrom in minmax:
                    minmax[chrom]=[]
                bedintervals[sample][chrom].append([s,e])
                minmax[chrom].append(s)
                minmax[chrom].append(e)
    minmaxval={}
    for chrom, se in minmax.items():
        minmaxval[chrom]=[np.min(se), np.max(se)]
    
    mat=[]
    samples=[]
    for sample, chrom_intervals in bedintervals.items():
        print(sample)
        tmp=[]
        for chrom, intervals in chrom_intervals.items():
            tmp2=[0 for _ in range(minmaxval[chrom][0], minmaxval[chrom][1],step)]
            for s, e in intervals:
                s=s-minmaxval[chrom][0]
                e=e-minmaxval[chrom][0]
                for i in range(s, e,step):
                    tmp2[i//step]=1
            tmp.extend(tmp2)
        mat.append(tmp)
        samples.append(sample)
    mat=np.array(mat)
    mat = mat[:, np.any(mat, axis=0)]
    print(mat[:100,:100])
    if method==&#34;pearson&#34;:
        #dmat=Parallel(n_jobs=-1)(delayed(_calc_pearson)(ind, mat) for ind in list(it.combinations(range(mat.shape[0]), 2)))
        dmat=np.corrcoef(mat)
        dmat=np.array(dmat)
        #dmat=squareform(dmat)
        print(dmat)
        #dmat+=np.identity(dmat.shape[0])
    else:
        dmat=squareform(pdist(mat, method))
    
            
    
    g=sns.clustermap(data=dmat,xticklabels=samples,yticklabels=samples,
               method=&#34;ward&#34;, cmap=palette,
               col_cluster=True,
               row_cluster=True,
               figsize=figsize,
               rasterized=True,
               #cbar_kws={&#34;label&#34;:&#34;Pearson correlation&#34;}, 
               annot=show_val,
               **clustermap_param)
    if method==&#34;pearson&#34;:
        title=&#34;Pearson correlation&#34;
    else:
        title=method+&#34; distance&#34;
    g.cax.set_ylabel(title, rotation=-90,va=&#34;bottom&#34;)
    plt.setp(g.ax_heatmap.get_yticklabels(), rotation=0)  # For y axis
    plt.setp(g.ax_heatmap.get_xticklabels(), rotation=90) # For x axis
    if show:
        plt.show()
        
    return {&#34;correlation&#34;: dmat,&#34;samples&#34;:samples}</code></pre>
</details>
</dd>
<dt id="omniplot.chipseq.plot_bigwig"><code class="name flex">
<span>def <span class="ident">plot_bigwig</span></span>(<span>files: dict, bed: Union[str, list], gff: str, step: int = 100, stack_regions: str = 'horizontal', highlight: Union[str, list] = [], highlightname='', n_jobs=-1, save: str = '') ‑> dict</span>
</code></dt>
<dd>
<div class="desc"><p>Plotting bigwig files in specified genomic regions in the bed file.
</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>files</code></strong> :&ensp;<code>dict</code></dt>
<dd>A dictionary whose keys are sample names and values are file names</dd>
<dt><strong><code>bed</code></strong> :&ensp;<code>str</code></dt>
<dd>A bed file name containing the list of genomic regions</dd>
<dt><strong><code>gff</code></strong> :&ensp;<code>str</code></dt>
<dd>A gff3 file with a corresponding genome version</dd>
<dt><strong><code>step</code></strong> :&ensp;<code>int</code></dt>
<dd>A bin size to reduce the bigwig signal resolution.</dd>
<dt><strong><code>show</code></strong> :&ensp;<code>bool</code></dt>
<dd>Whether or not to show the figure.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>{"ax":axes,"values":mat,"genes":geneset,"positions":pos} : dict
A dictionary containing matplot axes, signal values, geneset, and positions.
Raises</p>
<hr>
<h2 id="notes">Notes</h2>
<h2 id="references">References</h2>
<h2 id="see-also">See Also</h2>
<h2 id="examples">Examples</h2></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_bigwig(files: dict, 
                bed: Union[str, list], 
                gff: str,
                step: int=100,
                stack_regions: str=&#34;horizontal&#34;, 
                highlight: Union[str, list]=[],
                highlightname=&#34;&#34;,
                n_jobs=-1,
                save: str=&#34;&#34;)-&gt;dict:
    &#34;&#34;&#34;
    Plotting bigwig files in specified genomic regions in the bed file.  
    
    Parameters
    ----------
    files : dict
        A dictionary whose keys are sample names and values are file names 
    bed : str
        A bed file name containing the list of genomic regions
    gff : str
        A gff3 file with a corresponding genome version
    step: int
        A bin size to reduce the bigwig signal resolution.
    show : bool
        Whether or not to show the figure.
    Returns
    -------
    {&#34;ax&#34;:axes,&#34;values&#34;:mat,&#34;genes&#34;:geneset,&#34;positions&#34;:pos} : dict
        A dictionary containing matplot axes, signal values, geneset, and positions.
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34; 
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
    if type(bed)==str:
        posrange=[]
        pos=[]
        with open(bed) as fin:
            
            for l in fin:
                l=l.split()
                chrom, s, e=l[0],l[1],l[2]
                posrange.append(pr.from_dict({&#34;Chromosome&#34;: [chrom], &#34;Start&#34;: [int(s.replace(&#34;,&#34;,&#34;&#34;))], &#34;End&#34;: [int(e.replace(&#34;,&#34;,&#34;&#34;))]}))
                pos.append([chrom,int(s.replace(&#34;,&#34;,&#34;&#34;)),int(e.replace(&#34;,&#34;,&#34;&#34;))])  
    else:
        pos=bed
        
    if type(highlight)==str:
        _highlight=[]
        with open(bed) as fin:
            for l in fin:
                l=l.split()
                chrom, s, e=l[0],l[1],l[2]
                _highlight.append([chrom,int(s.replace(&#34;,&#34;,&#34;&#34;)),int(e.replace(&#34;,&#34;,&#34;&#34;))])
        highlight=_highlight
    
    time_start=time.time()
    if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
        gff_dict=_readgff2(gff, &#34;gene&#34;, others=[&#34;gene_name&#34;, &#34;Strand&#34;])
    elif gff.endswith(&#34;gtf&#34;):
        gff_dict=_readgtf2(gff, &#34;transcript&#34;, others=[&#34;gene_name&#34;, &#34;Strand&#34;])
    else:
        raise Exception(&#34;The gff/gtf file name is required to have either gff, gff3, or gtf extension.&#34;)
    gff_ob=pr.from_dict(gff_dict)
    
    #
    # if gff.endswith(&#34;gff&#34;) or gff.endswith(&#34;gff3&#34;):
    #     gff_ob=pr.read_gff3(gff)
    #     gff_ob=gff_ob[gff_ob.Feature==&#34;gene&#34;]
    #     gffformat=&#34;gff&#34;
    # elif gff.endswith(&#34;gtf&#34;):
    #     gff_ob=pr.read_gtf(gff)
    #     gff_ob=gff_ob[gff_ob.Feature==&#34;transcript&#34;]
    #     gffformat=&#34;gtf&#34;
    print(&#34;Reading the gene annotation file took&#34;, time.time()-time_start)
    # geneset=[]
    # for chrom, start, end in pos:
    #     geneset.append(gff_ob.get_genes(chrom, start, end))
    time_start=time.time()
    geneset=[]
    _grs=Parallel(n_jobs=n_jobs)(delayed(gff_ob.overlap)(gr) for gr in posrange)
    # for gr in posrange:
    #     gr2=gff_ob.intersect(gr)
    for gr2 in _grs:
        #gr2=gff_ob.intersect(gr)
        if len(gr2)==0:
            geneset.append([])
        else:
            #print(gr2)
            # if gffformat==&#34;gff&#34;:
            #     gr2=gr2[gr2.Feature==&#34;gene&#34;]
            # elif gffformat==&#34;gtf&#34;:
            #     gr2=gr2[gr2.Feature==&#34;transcript&#34;]
                
            tmp=[]
            for gene_name, start, end, ori in zip(gr2.gene_name, gr2.Start, gr2.End, gr2.Strand):
                tmp.append([gene_name, start, end, ori])
            geneset.append(tmp)
    #print(geneset)
    #geneset=Parallel(n_jobs=-1)(delayed(gff_ob.get_genes)(chrom, start, end) for chrom, start, end in pos)
    print(&#34;Obtaining genes took&#34;, time.time()-time_start)
            
    time_start=time.time()
    mat={}
    for samplename, f in files.items():
        bw=pwg.open(f)
        #h, t=os.path.split(f)
        #samplename=t.replace(prefix, &#34;&#34;).replace(suffix_set, &#34;&#34;)
        mat[samplename]=[]
        
        for chrom, s, e in pos:
            val=bw.values(chrom, s, e)
            
            mat[samplename].append(val)
    
    
    
    print(&#34;Reading bigwig files took&#34;, time.time()-time_start)
    
    if stack_regions==&#34;vertical&#34;:
        fig, axes=plt.subplots(nrows=len(pos)*2,ncols=len(files),figsize=[10,10],gridspec_kw={
                           &#39;height_ratios&#39;: [2,1]*len(pos),&#34;hspace&#34;:0},
                       constrained_layout = True)
        if len(axes.shape)==1:
            axes=axes.reshape([-1,1])
        sample_order=sorted(mat.keys())
        ticks_labels=[]
        for si, sample in enumerate(sample_order):
            
            vals=mat[sample]
            ymax=0
            for posi, (genes, val) in enumerate(zip(geneset, vals)):
                if len(val)==0:
                    continue
                ax=axes[posi*2,si]
                chrom, s, e=pos[posi]
                val=np.array(val)
                vallen=val.shape[0]
                remains=vallen%step
                _e=e-remains
                x=np.arange(s, _e,step)
                #print(s, e, _e, vallen, step)
                val=val[:vallen-remains].reshape(-1, step).mean(axis=1)
                _ymax=np.amax(val)
                if _ymax &gt;  ymax:
                    ymax =_ymax
                if len(highlight)!=0:
                    hchrom, hs, he=highlight[posi]
                    ax.fill_between([hs, he], [_ymax, _ymax], alpha=0.5, color=&#34;r&#34;)
                ax.fill_between(x, val,edgecolor=&#39;face&#39;)
                
                
                if si==0:
                    ax.set_title(sample+&#34; &#34;+chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e))
                    ticks_labels.append([[np.amin(x),np.amax(x)],ax.get_xticks(), ax.get_xticklabels()])
                ax.set_xticks([],labels=[])
            occupied=[[] for i in range(8)]
        
        

        
        for posi in range(len(pos)):
            interval=pos[posi][2]-pos[posi][1]
            axes[posi*2+1,si].plot(ticks_labels[posi][0],[0,0],alpha=0)
            for genename, gs, ge,ori in geneset[posi]:
                slot=0
                for oi in range(len(occupied)):
                    oc=False
                    if len(occupied[oi])==0:
                        occupied[oi].append([gs, ge])
                        oc=False
                        slot=oi
                    else:
                        for tmps, tmpe in occupied[oi]:
                            if tmps&lt;=gs&lt;= tmpe or tmps&lt;=ge&lt;=tmpe:
                                oc=True
                        if oc==False:
                            occupied[oi].append([gs, ge])
                            slot=oi
                    if oc==False:
                        break
                if gs &lt; pos[posi][1]:
                    gs=pos[posi][1]
                if ge &gt; pos[posi][2]:
                    ge=pos[posi][2]
                axes[posi*2+1,si].plot([gs, ge], [slot*1,slot*1], color=&#34;gray&#34;)
                if ori==&#34;+&#34;:
                    axes[posi*2+1,si].plot([ge-interval/32, ge], [slot*1+1,slot*1], color=&#34;gray&#34;)
                elif ori==&#34;-&#34;:
                    axes[posi*2+1,si].plot([gs, gs+interval/32], [slot*1,slot*1+1], color=&#34;gray&#34;)
                
                axes[posi*2+1,si].text((gs+ge)/2, slot*1, genename, ha=&#34;center&#34;)
            for oi, ol in enumerate(occupied):
                if len(ol)==0:
                    maxslot=oi
                    break
            axes[posi*2+1,si].set_ylim(-0.5,maxslot+0.5)
            #axes[posi*2+1,si].ticklabel_format(useOffset=False)
            axes[posi*2+1,si].set_yticks([],labels=[])
            axes[posi*2+1,si].set_xticks(ticks_labels[posi][1])
            axes[posi*2+1,si].set_ylabel(&#34;Genes&#34;)
        if save !=&#34;&#34;:
            if save.endswith(&#34;.pdf&#34;) or save.endswith(&#34;.png&#34;) or save.endswith(&#34;.svg&#34;):
                h, t=os.path.splitext(save)
                plt.savefig(h+&#34;_plot&#34;+t)
            else:
                plt.savefig(save+&#34;_plot.pdf&#34;)
    else:
        fig, axes=plt.subplots(nrows=len(files)+1,ncols=len(pos),gridspec_kw={
                           &#39;height_ratios&#39;: [2]*len(files)+[1]})
        
        if len(axes.shape)==1:
            axes=axes.reshape([-1,1])
        sample_order=sorted(mat.keys())
        for si, sample in enumerate(sample_order):
            vals=mat[sample]
            ymax=0
            #print(ymax)
            for posi, (genes, val) in enumerate(zip(geneset, vals)):
                if len(val)==0:
                    continue
                ax=axes[si,posi]
                chrom, s, e=pos[posi]
                val=np.array(val)
                vallen=val.shape[0]
                remains=vallen%step
                _e=e-remains
                x=np.arange(s, _e,step)
                
                #print(s, e, _e, vallen, step)
                val=val[:vallen-remains].reshape(-1, step).mean(axis=1)
                _ymax=np.amax(val)
                if _ymax &gt;  ymax:
                    ymax =_ymax

                ax.fill_between(x, val,edgecolor=&#39;face&#39;)

                if posi==0:
                    ax.set_ylabel(sample)
    
                ax.set_xticks([],labels=[])
                if si==0:
                    ax.set_title(chrom+&#34;:&#34;+str(s)+&#34;-&#34;+str(e))
                ax.grid(False)
                    
            occupied=[[] for i in range(8)]
            for posi in range(len(pos)):
                axes[si,posi].set_ylim(0,ymax*1.01)
                if posi &gt;0:
                    axes[si,posi].set_yticks([],labels=[])
        
        for posi in range(len(pos)):
            interval=pos[posi][2]-pos[posi][1]
            for genename, gs, ge,ori in geneset[posi]:
                slot=0
                for oi in range(len(occupied)):
                    oc=False
                    if len(occupied[oi])==0:
                        occupied[oi].append([gs, ge])
                        oc=False
                        slot=oi
                    else:
                        for tmps, tmpe in occupied[oi]:
                            if tmps&lt;=gs&lt;= tmpe or tmps&lt;=ge&lt;=tmpe:
                                oc=True
                        if oc==False:
                            occupied[oi].append([gs, ge])
                            slot=oi
                    if oc==False:
                        break
                if gs &lt; pos[posi][1]:
                    gs=pos[posi][1]
                if ge &gt; pos[posi][2]:
                    ge=pos[posi][2]
                axes[si+1,posi].plot([gs, ge], [slot*1,slot*1], color=&#34;gray&#34;)
                if ori==&#34;+&#34;:
                    axes[si+1,posi].plot([ge-interval/32, ge], [slot*1+0.1,slot*1], color=&#34;gray&#34;)
                elif ori==&#34;-&#34;:
                    axes[si+1,posi].plot([gs, gs+interval/32], [slot*1,slot*1+0.1], color=&#34;gray&#34;)
                
                axes[si+1,posi].text((gs+ge)/2, slot*1, genename, ha=&#34;center&#34;)
                #axes[si+1,posi].grid(False)
            for oi, ol in enumerate(occupied):
                if len(ol)==0:
                    maxslot=oi
                    break
            axes[si+1,posi].set_ylim(-0.5,maxslot+0.5)
            axes[si+1,posi].ticklabel_format(useOffset=False)
            axes[si+1,posi].set_yticks([],labels=[])
            
        
        plt.tight_layout(h_pad=-1,w_pad=-1)
        _save(save, &#34;plot&#34;)
        
    return {&#34;ax&#34;:axes,&#34;values&#34;:mat,&#34;genes&#34;:geneset,&#34;positions&#34;:pos}</code></pre>
</details>
</dd>
<dt id="omniplot.chipseq.plot_bigwig_correlation"><code class="name flex">
<span>def <span class="ident">plot_bigwig_correlation</span></span>(<span>files: dict, chrom: str = 'chr1', step: int = 1000, palette: str = 'coolwarm', figsize: list = [6, 6], show_val: bool = True, clustermap_param: dict = {}, peakfile: str = '', show: bool = False, n_jobs: int = -1, save: str = '') ‑> Dict[~KT, ~VT]</span>
</code></dt>
<dd>
<div class="desc"><p>Calculate pearson correlations and draw a heatmap for bigwig files.
</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>files</code></strong> :&ensp;<code>dict</code></dt>
<dd>A dictionary whose keys are sample names and values are bigwig file names</dd>
<dt><strong><code>chrom</code></strong> :&ensp;<code>str</code>, optional</dt>
<dd>A chromsome to cal</dd>
<dt><strong><code>step</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>A bin size to reduce the bigwig signal resolution.</dd>
<dt><strong><code>palette</code></strong> :&ensp;<code>str</code>, optional</dt>
<dd>&nbsp;</dd>
<dt><strong><code>show_val</code></strong> :&ensp;<code>bool</code>, optional</dt>
<dd>Whether or not to show correlation values on the heatmap</dd>
<dt><strong><code>peakfile</code></strong> :&ensp;<code>str</code>, optional</dt>
<dd>A peak file/bed file containing genome regions of interest. If given this option, the correlation will be calculated using the genome regions listed in this file.</dd>
<dt><strong><code>show</code></strong> :&ensp;<code>bool</code>, optional</dt>
<dd>Whether or not to show the figure.</dd>
<dt><strong><code>n_jobs</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>The number of threads to use. It will use all available threads by default.</dd>
</dl>
<h2 id="returns">Returns</h2>
<p>{"ax":axes,"values":mat,"genes":geneset,"positions":pos} : dict
A dictionary containing matplot axes, signal values, geneset, and positions.
Raises</p>
<hr>
<h2 id="notes">Notes</h2>
<h2 id="references">References</h2>
<h2 id="see-also">See Also</h2>
<h2 id="examples">Examples</h2></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def plot_bigwig_correlation(files: dict, 
                            chrom: str=&#34;chr1&#34;,
                            step: int=1000,
                            palette: str=&#34;coolwarm&#34;,
                            figsize: list=[6,6],
                            show_val: bool=True,
                            clustermap_param: dict={},
                            peakfile: str=&#34;&#34;,
                            #show_val: bool=False,
                            show: bool=False,
                            n_jobs: int=-1, save: str=&#34;&#34;) -&gt; Dict:
    
    &#34;&#34;&#34;
    Calculate pearson correlations and draw a heatmap for bigwig files.  
    
    Parameters
    ----------
    files : dict
        A dictionary whose keys are sample names and values are bigwig file names 
    chrom : str, optional
        A chromsome to cal
    step: int, optional
        A bin size to reduce the bigwig signal resolution.
    palette: str, optional
    show_val: bool, optional
        Whether or not to show correlation values on the heatmap
    peakfile: str, optional
        A peak file/bed file containing genome regions of interest. If given this option, the correlation will be calculated using the genome regions listed in this file.
    show : bool, optional
        Whether or not to show the figure.
    n_jobs : int, optional
        The number of threads to use. It will use all available threads by default. 
    Returns
    -------
    {&#34;ax&#34;:axes,&#34;values&#34;:mat,&#34;genes&#34;:geneset,&#34;positions&#34;:pos} : dict
        A dictionary containing matplot axes, signal values, geneset, and positions.
    Raises
    ------
    Notes
    -----
    References
    ----------
    See Also
    --------
    Examples
    --------
    &#34;&#34;&#34;
    for k, v in files.items():
        assert type(v)==str, &#34;filenames must be str, but {} was give.&#34;.format(type(v))
        if not os.path.isfile(v)==True:
            raise Exception(&#34;{} does not exist.&#34;.format(v))
    
    mat=[]
    samples=[]
    bw=pwg.open(list(files.values())[0])
    chrom_sizes=bw.chroms()
    
    
        
    
    if peakfile!=&#34;&#34;:
        # peaks=read_peaks(peakfile)
        # for k, f in files.items():
        #     bw=pwg.open(f)
        #     tmp=[]
        #     for chrom, se in peaks.items():
        #         for s, e in se:
        #             val=bw.stats(chrom,s,e, exact=True)[0]
        #             if val==None:
        #                 val=0
        #         tmp.append(val)
        #     mat.append(tmp)
        #     samples.append(k)
        peaks=_read_peaks(peakfile)
        for k, f in files.items():
            bw=pwg.open(f)
            
            tmp=Parallel(n_jobs=n_jobs)(delayed(_read_bw_stats)(chrom, s, e, f) for chrom, s, e in peaks)
            mat.append(tmp)
            samples.append(k)
        
    elif chrom !=&#34;all&#34;:
        chrom_size=chrom_sizes[chrom]
        
        samples=[]
        mat=[]
        #samples=list(files.keys())
        #mat=Parallel(n_jobs=n_jobs)(delayed(read_bw)(files[s],chrom, chrom_size, step) for s in samples)
        for k, f in files.items():
            bw=pwg.open(f)
            val=np.array(bw.values(chrom,0,chrom_size))
            rem=chrom_size%step
            val=val[:val.shape[0]-rem]
            print(val.shape)
            val=np.nan_to_num(val)
            val=val.reshape([-1,step]).mean(axis=1)
            print(val.shape)
            mat.append(val/np.sum(val))
            samples.append(k)
    elif chrom==&#34;all&#34;:
        chroms=chrom_sizes.keys()
        for k, f in files.items():
            
            #tmp=Parallel(n_jobs=n_jobs)(delayed(read_bw)(f, chrom, 0, chrom_sizes[chrom], f) for chrom, s, e in chroms)
            bw=pwg.open(f)
            tmp=[]
            for chrom in chroms:
                chrom_size=chrom_sizes[chrom]
                val=np.array(bw.values(chrom,0,chrom_size))
                rem=chrom_size%step
                val=val[:val.shape[0]-rem]
                val=np.nan_to_num(val)
                val=val.reshape([-1,step]).mean(axis=1)
                tmp.append(val/np.sum(val))
            
            samples.append(k)
            mat.append(np.concatenate(tmp))
    else:
        raise Exception(&#34;please specify chrom or peakfile options.&#34;)
    
    mat=np.array(mat)
    dmat=Parallel(n_jobs=-1)(delayed(_calc_pearson)(ind, mat) for ind in list(it.combinations(range(mat.shape[0]), 2)))
    dmat=np.array(dmat)
    get_reusable_executor().shutdown(wait=True)
    dmat=squareform(dmat)
    print(dmat)
    dmat+=np.identity(dmat.shape[0])
    g=sns.clustermap(data=dmat,xticklabels=samples,yticklabels=samples,
               method=&#34;ward&#34;, cmap=palette,
               col_cluster=True,
               row_cluster=True,
               figsize=figsize,
               rasterized=True,
               #cbar_kws={&#34;label&#34;:&#34;Pearson correlation&#34;}, 
               annot=show_val,
               **clustermap_param)
    g.cax.set_ylabel(&#34;Pearson correlation&#34;, rotation=-90,va=&#34;bottom&#34;)
    plt.setp(g.ax_heatmap.get_yticklabels(), rotation=0)  # For y axis
    plt.setp(g.ax_heatmap.get_xticklabels(), rotation=90) # For x axis
    
    
    _save(save, &#34;correlation&#34;)
    if show:
        plt.show()
    return {&#34;correlation&#34;: dmat,&#34;samples&#34;:samples}</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="omniplot" href="index.html">omniplot</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="omniplot.chipseq.call_superenhancer" href="#omniplot.chipseq.call_superenhancer">call_superenhancer</a></code></li>
<li><code><a title="omniplot.chipseq.frip_calc" href="#omniplot.chipseq.frip_calc">frip_calc</a></code></li>
<li><code><a title="omniplot.chipseq.plot_average" href="#omniplot.chipseq.plot_average">plot_average</a></code></li>
<li><code><a title="omniplot.chipseq.plot_bed_correlation" href="#omniplot.chipseq.plot_bed_correlation">plot_bed_correlation</a></code></li>
<li><code><a title="omniplot.chipseq.plot_bigwig" href="#omniplot.chipseq.plot_bigwig">plot_bigwig</a></code></li>
<li><code><a title="omniplot.chipseq.plot_bigwig_correlation" href="#omniplot.chipseq.plot_bigwig_correlation">plot_bigwig_correlation</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>