<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>omniplot.chipseq_utils API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>omniplot.chipseq_utils</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from typing import List,Dict,Optional,Union, Any, Iterable
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
try:
    import pyBigWig as pwg
except ImportError:
    pass
import time
import os 
import scipy.stats
from scipy.spatial.distance import squareform
from scipy.spatial import distance
from joblib import Parallel, delayed
import itertools as it
from joblib.externals.loky import get_reusable_executor
sns.set_theme(font=&#34;Arial&#34;, style={&#39;grid.linestyle&#39;: &#34;&#34;,&#39;axes.facecolor&#39;: &#39;white&#39;})
import itertools
from sklearn.neighbors import BallTree
from intervaltree import Interval, IntervalTree
import pickle
import os

#from cython_utils import chipseq_utils
def _range_diff(r1, r2):
    s1, e1 = r1
    s2, e2 = r2
    endpoints = sorted((s1, s2, e1, e2))
    result = []
    if endpoints[0] == s1 and endpoints[1] != s1:
        result.append((endpoints[0], endpoints[1]))
    if endpoints[3] == e1 and endpoints[2] != e1:
        result.append((endpoints[2], endpoints[3]))
    return result

def _multirange_diff(r1_list, r2_list):
    for r2 in r2_list:
        r1_list = list(itertools.chain(*[_range_diff(r1, r2) for r1 in r1_list]))
    return r1_list

class gff_parser():
    resolution=100000
    def __init__(self,gff):
        from operator import methodcaller
        time_start=time.time()
        home=os.path.expanduser(&#39;~&#39;)
        cache=os.path.join(home, &#34;omniplot_cache&#34;)
        if not os.path.isdir(cache):
            os.makedirs(cache)
        h, t=os.path.split(gff)
        gffchache=os.path.join(cache, t+&#34;.pickle&#34;)
        gffchache_info=os.path.join(cache, t+&#34;_info.pickle&#34;)
        if os.path.isfile(gffchache):
            with open(gffchache, &#34;rb&#34;) as fin:
                data=pickle.load(fin)
            with open(gffchache_info, &#34;rb&#34;) as fin:
                info=pickle.load(fin)
        else:
            data={}
            info={}
            with open(gff) as fin:
                for l in fin:
                    if l.startswith(&#34;#&#34;):
                        if l.startswith(&#34;##gff-version&#34;):
                            l=l.split()
                            info[&#34;gff-version&#34;]=l[1]
                        elif l.startswith(&#34;#description:&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;description&#34;]=l[1]
                        elif l.startswith(&#34;#provider:&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;provider&#34;]=l[1]
                        elif l.startswith(&#34;#format&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;format&#34;]=l[1]
                        elif l.startswith(&#34;#date&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;date&#34;]=l[1]
                        continue
                    chrom, source, kind, s, e, _, ori, _, meta=l.split()
                    s, e=int(s)-1, int(e)
                    meta=meta.split(&#34;;&#34;)
                    tmp={}
                    for m in meta:
                        k, v=m.split(&#34;=&#34;)
                        tmp[k]=v
                    

                    # if not chrom in data:
                    #     data[chrom]={&#34;pos&#34;:[], &#34;kind&#34;:[],&#34;ori&#34;:[],&#34;meta&#34;:[]}
                    #
                    #
                    # #tmp={k: v for k,v in map(methodcaller(&#34;split&#34;, &#34;=&#34;), meta)}
                    # data[chrom][&#34;pos&#34;].append([s, e])
                    # data[chrom][&#34;kind&#34;].append(kind)
                    # data[chrom][&#34;ori&#34;].append(ori)
                    # data[chrom][&#34;meta&#34;].append(tmp)        
                    if not chrom in data:
                        data[chrom]={}
                  
                    _s=self.resolution*(s//self.resolution)
                    if not _s in data[chrom]:
                        data[chrom][_s]={&#34;pos&#34;:[], &#34;kind&#34;:[],&#34;ori&#34;:[],&#34;meta&#34;:[]}
                    data[chrom][_s][&#34;pos&#34;].append([s, e])
                    data[chrom][_s][&#34;kind&#34;].append(kind)
                    data[chrom][_s][&#34;ori&#34;].append(ori)
                    data[chrom][_s][&#34;meta&#34;].append(tmp)  
            with open(gffchache, &#34;wb&#34;) as fin:
                pickle.dump(data, fin)
            with open(gffchache_info, &#34;wb&#34;) as fin:
                pickle.dump(info, fin)         
        self.data=data
        self.info=info
        print(time.time()-time_start)
    def get_genes(self, chrom: str,
                  start: int, 
                  end: int,
                  gene_type: set=set([&#34;protein_coding&#34;])) -&gt; list:
        _data=self.data[chrom]
        genes=[]
        
        
        _start=start//self.resolution
        _end=end//self.resolution
        for h in range(_start, _end+1):
            _h=self.resolution*h
            _data2=_data[_h]
            for i, (se, kind, meta, ori) in enumerate(zip(_data2[&#34;pos&#34;], _data2[&#34;kind&#34;], _data2[&#34;meta&#34;], _data2[&#34;ori&#34;])):
                s, e=se
                if e &lt; start:
                    continue
                if s&gt;end:
                    break
                if kind==&#34;gene&#34; and meta[&#34;gene_type&#34;] in gene_type:
                    
                
                    if start&lt;=s and e&lt;=end:
                        genename=meta[&#34;gene_name&#34;]
                        genes.append([genename,s,e,ori])
                    elif start&lt;=s&lt;=end and end&lt;e:
                        genename=meta[&#34;gene_name&#34;]
                        genes.append([genename,s,end,ori])
                    elif s&lt;start and start&lt;=e&lt;=end  and meta[&#34;gene_type&#34;] in gene_type:
                        genename=meta[&#34;gene_name&#34;]
                        genes.append([genename,start,e,ori])
        
        
        # _data=self.data[chrom]
        # genes=[]
        # for i, (se, kind, meta, ori) in enumerate(zip(_data[&#34;pos&#34;], _data[&#34;kind&#34;], _data[&#34;meta&#34;], _data[&#34;ori&#34;])):
        #     s, e=se
        #     if e &lt; start:
        #         continue
        #     if s&gt;end:
        #         break
        #     if kind==&#34;gene&#34; and meta[&#34;gene_type&#34;] in gene_type:
        #
        #
        #         if start&lt;=s and e&lt;=end:
        #             genename=meta[&#34;gene_name&#34;]
        #             genes.append([genename,s,e,ori])
        #         elif start&lt;=s&lt;=end and end&lt;e:
        #             genename=meta[&#34;gene_name&#34;]
        #             genes.append([genename,s,end,ori])
        #         elif s&lt;start and start&lt;=e&lt;=end  and meta[&#34;gene_type&#34;] in gene_type:
        #             genename=meta[&#34;gene_name&#34;]
        #             genes.append([genename,start,e,ori])
                
            

        return genes
    def get_tss(self, gene_type=set([&#34;protein_coding&#34;])):
        
        tss={}
        
        for chrom, _data in self.data.items():
            tss[chrom]={&#34;pos&#34;:[],&#34;genes&#34;:[]}
            for i, (s,e) in enumerate(_data[&#34;pos&#34;]):
                if _data[&#34;kind&#34;][i]==&#34;gene&#34; and _data[&#34;meta&#34;][i][&#34;gene_type&#34;] in gene_type:
                    tss[chrom][&#34;genes&#34;].append(_data[&#34;meta&#34;][i][&#34;gene_name&#34;])
                    if _data[&#34;ori&#34;][i]==&#34;+&#34;:
                        tss[chrom][&#34;pos&#34;].append(s)
                    else:
                        tss[chrom][&#34;pos&#34;].append(e)
        
        return tss 
    
    def get_tss_extend(self, gene_type=set([&#34;protein_coding&#34;]), extend=5000):
        
        tss={}
        
        for chrom, _data in self.data.items():
            tss[chrom]={&#34;pos&#34;:[],&#34;genes&#34;:[]}
            for i, (s,e) in enumerate(_data[&#34;pos&#34;]):
                if _data[&#34;kind&#34;][i]==&#34;gene&#34; and _data[&#34;meta&#34;][i][&#34;gene_type&#34;] in gene_type:
                    tss[chrom][&#34;genes&#34;].append(_data[&#34;meta&#34;][i][&#34;gene_name&#34;])
                    if _data[&#34;ori&#34;][i]==&#34;+&#34;:
                        tss[chrom][&#34;pos&#34;].append([s-extend,s+extend])
                    else:
                        tss[chrom][&#34;pos&#34;].append([e-extend,e+extend])
        
        return tss
    
    def closest_genes(self, peaks, K=1):
        tss=self.get_tss()
        
        if type(peaks)==list:
            _peaks={}
            for chrom, s, e in peaks:
                if not chrom in _peaks:
                    _peaks[chrom]={&#34;se&#34;:[],&#34;center&#34;:[]}
                _peaks[chrom][&#34;center&#34;].append((s+e)/2)
                _peaks[chrom][&#34;se&#34;].append([s,e])
            peaks=_peaks
        genes_peaks=[]
        i=0
        for chrom in tss.keys():
            if not chrom in peaks:
                continue
            _tss=np.array(tss[chrom][&#34;pos&#34;]).reshape([-1,1])
            _genes=np.array(tss[chrom][&#34;genes&#34;])
            _peaks=np.array(peaks[chrom][&#34;center&#34;]).reshape([-1,1])
            tree = BallTree(_tss)
            dist, ind = tree.query(_peaks, k=K) 
            for _ind, _dist, (s,e) in zip(list(ind), list(dist),peaks[chrom][&#34;se&#34;]):
                _closest=_genes[_ind]
                genes_peaks.append([chrom, s, e, &#34;,&#34;.join(list(_closest)), &#34;,&#34;.join(map(str, list(_dist)))])
        return genes_peaks
    
def _calc_pearson2(_ind, _mat):
    _a, _b=_mat[_ind[0]], _mat[_ind[1]]
    _af=_a&gt;np.quantile(_a, 0.75)
    _bf=_b&gt;np.quantile(_b, 0.75)
    #_ab=
    #filt=_ab&gt;0.02
    filt=_af+_bf
    _a=_a[filt]
    _b=_b[filt]
    p=scipy.stats.pearsonr(_a, _b)[0]
    return p
def _calc_pearson(_ind, _mat):
    p=scipy.stats.pearsonr(_mat[_ind[0]], _mat[_ind[1]])[0]
    #p=scipy.stats.spearmanr(_mat[_ind[0]], _mat[_ind[1]])[0]
    #p=distance.cdist([_mat[_ind[0]]], [_mat[_ind[1]]])[0][0]
    return p

def _read_bw(f, chrom, chrom_size, step):
        bw=pwg.open(f)
        val=np.array(bw.values(chrom,0,chrom_size))
        rem=chrom_size%step
        val=val[:val.shape[0]-rem]
        val=np.nan_to_num(val)
        val=val.reshape([-1,step]).mean(axis=1)
        return val
def _read_peaks(peakfile):
    peaks={}
    with open(peakfile) as fin:
        for l in fin:
            l=l.split()
            if not l[0] in peaks:
                peaks[l[0]]=[]
            peaks[l[0]].append([int(l[1]),int(l[2])])
    return peaks

def _stitching(peakfile, stitchdist):
    speaks={}
    
    for chrom, intervals in peakfile.items():
        intervals.sort()
        stack = []
        # insert first interval into stack
        stack.append(intervals[0])
        for i in intervals[1:]:
            # Check for overlapping interval,
            # if interval overlap
            if i[0] - stack[-1][-1] &lt;stitchdist:
                stack[-1][-1] = max(stack[-1][-1], i[-1])
            else:
                stack.append(i)
        speaks[chrom]=stack
    return speaks

def _stitching_for_pyrange(peakfile, stitchdist):
    speaks={&#34;Chromosome&#34;:[],&#34;Start&#34;:[],&#34;End&#34;:[]}
    
    for chrom, intervals in peakfile.items():
        intervals.sort()
        stack = []
        # insert first interval into stack
        stack.append(intervals[0])
        for i in intervals[1:]:
            # Check for overlapping interval,
            # if interval overlap
            if i[0] - stack[-1][-1] &lt;stitchdist:
                stack[-1][-1] = max(stack[-1][-1], i[-1])
            else:
                stack.append(i)
        for s, e in stack:
            speaks[&#34;Chromosome&#34;].append(chrom)
            speaks[&#34;Start&#34;].append(s)
            speaks[&#34;End&#34;].append(e)
        #speaks[chrom]=stack
    return speaks

def _read_tss(tss, tss_dist):
    tss_pos={}
    with open(tss) as fin:
        for l in fin:
            l=l.split()
            if not l[0] in tss_pos:
                tss_pos[l[0]]=[]
            if l[4]==&#34;+&#34;:
                tss_pos[l[0]].append([int(l[1])-tss_dist,int(l[1])+tss_dist])
            elif l[4]==&#34;-&#34;:
                tss_pos[l[0]].append([int(l[2])-tss_dist,int(l[2])+tss_dist])
            
            else:
                raise Exception(&#34;TSS bed file format must look like &#39;chromsome\tstart\tend\tname\torientation&#39;&#34;)
            
    return tss_pos

def _interval_subtraction(list1, list2):
    t = IntervalTree([Interval(s, e) for s, e in list1])

    newt=Parallel(n_jobs=-1)(delayed(t.chop)(s, e) for s, e in list2)
    #print(newt[:10])
    _newt=[]
    for val in newt:
        if val!=None:
            _newt.append(val)
    newt=sorted(_newt)
    return [[_t.begin, _t.end] for _t in newt]
def _remove_close_to_tss(stitched, tss_pos):
    _stitched={}
    chroms=stitched.keys()
    # tmp=[]
    # for chrom in chroms:
    #     tmp.extend(interval_subtraction(stitched[chrom], tss_pos[chrom]))
    # print(tmp[:10])
    tmp=Parallel(n_jobs=-1)(delayed(_multirange_diff)(stitched[chrom], tss_pos[chrom]) for chrom in chroms)
    _stitched={chrom: _tmp for chrom, _tmp in zip(chroms,tmp) }
    # for chrom, se in stitched.items():
    #
    #     tsslist=tss_pos[chrom]
    #     _stitched[chrom]=multirange_diff(se, tsslist)
        
    #
    # for s, e in se:
    #     news=[s]
    #     newe=[e]
    #     tss_inbetween=False
    #     for _tss, ori in tsslist:
    #         if s+tss_dist&lt;_tss:
    #             continue
    #         if _tss &gt;e+tss_dist:
    #             break
    #         if s &lt;= _tss &lt;= e:
    #             newe.append(_tss-tss_dist)
    #             news.append(_tss+tss_dist)
    #             tss_inbetween=True
    #             continue
    #         if 0&lt; _tss-e&lt; tss_dist:
    #             _e=_tss-tss_dist
    #             newe.append(_e)
    #         if 0 &lt; s-_tss &lt; tss_dist:
    #             _s=_tss+tss_dist
    #             news.append(_s)
    #     _e=min(newe)
    #     _s=max(news)
    #     if _s==107683009:
    #         print(chrom, s,e)
    #     if _e &lt;=s:
    #         continue
    #     if e &lt;=_s:
    #         continue
    #     if tss_inbetween:
    #         _stitched[chrom].append([_s,e])
    #         _stitched[chrom].append([s,_e])
    #     else:
    #         _stitched[chrom].append([_s,_e])
    return _stitched

def _find_extremes(signals, pos):
    signals=np.array(signals)
    srt=np.argsort(signals)
    signals=signals[srt]
    pos=np.array(pos)
    pos=pos[srt]
    
    y=signals
    _y=y/np.amax(y)
    x=np.linspace(0,1, y.shape[0])
    b=0
    possitives=[]
    for i in reversed(range(x.shape[0])):
        __y=x-i/x.shape[0] + _y[i]
        #print(y)
        possitives.append(np.sum(_y-__y&gt;0))

    index=len(possitives)-np.argmax(possitives)
    
    
    return x, y, pos, index, srt

def _readgff(_gff, _kind, tss_dist):
    gff_dict={&#34;Chromosome&#34;:[], &#34;Start&#34;: [], &#34;End&#34;: []}
    with open(_gff) as fin:
        for l in fin:
            if l.startswith(&#34;#&#34;):
                continue
            chrom, source, kind, s, e, _, ori, _, meta=l.split()
            if kind !=_kind:
                continue
            s, e=int(s)-1, int(e)
            if ori==&#34;+&#34;:
                gff_dict[&#34;Chromosome&#34;].append(chrom)
                gff_dict[&#34;Start&#34;].append(s-tss_dist)
                gff_dict[&#34;End&#34;].append(s+tss_dist)
            else:
                gff_dict[&#34;Chromosome&#34;].append(chrom)
                gff_dict[&#34;Start&#34;].append(e-tss_dist)
                gff_dict[&#34;End&#34;].append(e+tss_dist)
    return gff_dict

def _readgtf(_gff, _kind,tss_dist):
    gff_dict={&#34;Chromosome&#34;:[], &#34;Start&#34;: [], &#34;End&#34;: []}
    with open(_gff) as fin:
        for l in fin:
            if l.startswith(&#34;#&#34;):
                continue
            chrom, source, kind, s, e, _, ori, _, meta=l.strip(&#34;\n&#34;).split(&#34;\t&#34;)
            if kind !=_kind:
                continue
            s, e=int(s)-1, int(e)
            if ori==&#34;+&#34;:
                gff_dict[&#34;Chromosome&#34;].append(chrom)
                gff_dict[&#34;Start&#34;].append(s-tss_dist)
                gff_dict[&#34;End&#34;].append(s+tss_dist)
            else:
                gff_dict[&#34;Chromosome&#34;].append(chrom)
                gff_dict[&#34;Start&#34;].append(e-tss_dist)
                gff_dict[&#34;End&#34;].append(e+tss_dist)
    return gff_dict

def _readgff2(_gff, _kind, others=[]):
    gff_dict={&#34;Chromosome&#34;:[], &#34;Start&#34;: [], &#34;End&#34;: []}
    for other in others:
        gff_dict[other]=[]
    with open(_gff) as fin:
        for l in fin:
            if l.startswith(&#34;#&#34;):
                continue
            chrom, source, kind, s, e, _, ori, _, meta=l.split()
            if kind !=_kind:
                continue
            tmp={}
            #print(meta)
            meta=meta.split(&#34;;&#34;)
            
            for m in meta:
                k, v=m.split(&#34;=&#34;)
                if k in others:
                    tmp[k]=v
            s, e=int(s)-1, int(e)
            gff_dict[&#34;Chromosome&#34;].append(chrom)
            gff_dict[&#34;Start&#34;].append(s)
            gff_dict[&#34;End&#34;].append(e)
            for other in others:
                if other in tmp:
                    gff_dict[other].append(tmp[other])
                elif other==&#34;Strand&#34;:
                    gff_dict[other].append(ori)
    return gff_dict

def _readgff_transcripts(_gff, _extend, avoid=&#34;chrM&#34;):
    
    transcripts={&#34;+&#34;:{},&#34;-&#34;:{}}
    
    with open(_gff) as fin:
        for l in fin:
            if l.startswith(&#34;#&#34;):
                continue
            chrom, source, kind, s, e, _, ori, _, meta=l.split()
            if chrom==avoid:
                continue
            s, e=int(s)-1, int(e)
            if kind ==&#34;exon&#34;:
                tmp={}
                #print(meta)
                meta=meta.split(&#34;;&#34;)
                
                for m in meta:
                    k, v=m.split(&#34;=&#34;)
                    tmp[k]=v
                if &#34;pseudogene&#34; in tmp[&#34;gene_type&#34;].split(&#34;_&#34;):
                    continue
                tid=tmp[&#34;transcript_id&#34;]
                if not tid in transcripts[ori]:
                    transcripts[ori][tid]=[]
                transcripts[ori][tid].append([chrom, s, e])
    _transcripts={&#34;+&#34;:{},&#34;-&#34;:{}}
    for strand, tids in transcripts.items():
        if strand==&#34;+&#34;:
            for tid, exons in tids.items():
                _transcripts[strand][tid]=[]
                total_len=0
                i=0
                for chrom, s, e in exons:
            
                    total_len+=e-s
                    if i==0:
                        s=s-_extend
                        if s &lt; 0:
                            s=0
                    if total_len&gt;_extend:
                        e=e-(total_len-_extend)
                        if s&lt;e:
                            _transcripts[strand][tid].append([chrom, s, e])
                        break
                    else:
                        _transcripts[strand][tid].append([chrom, s, e])
                    i+=1
                
        elif strand==&#34;-&#34;:
            for tid, exons in tids.items():
                _transcripts[strand][tid]=[]
                total_len=0
                i=0
                for chrom, s, e in reversed(exons):
                    total_len+=e-s
                    if i==0:
                        e=e+_extend
                    if total_len&gt;_extend:
                        s=s+(total_len-_extend)
                        if s&lt;e:
                            _transcripts[strand][tid].append([chrom, s, e])
                        break
                    else:
                        _transcripts[strand][tid].append([chrom, s, e])
                    i+=1
    return _transcripts

def _readgtf2(_gff, _kind, others=[]):
    gff_dict={&#34;Chromosome&#34;:[], &#34;Start&#34;: [], &#34;End&#34;: []}
    for other in others:
        gff_dict[other]=[]
    with open(_gff) as fin:
        for l in fin:
            if l.startswith(&#34;#&#34;):
                continue
            chrom, source, kind, s, e, _, ori, _, meta=l.strip(&#34;;\n&#34;).split(&#34;\t&#34;)
            if kind !=_kind:
                continue
            tmp={}
            #print(meta)
            meta=meta.split(&#34;; &#34;)
            #print(meta)
            for m in meta:
                k, v=m.split()
                v=v.strip(&#39;&#34;;&#39;)
                if k in others:
                    tmp[k]=v

            s, e=int(s)-1, int(e)
            gff_dict[&#34;Chromosome&#34;].append(chrom)
            gff_dict[&#34;Start&#34;].append(s)
            gff_dict[&#34;End&#34;].append(e)
            for other in others:
                if other in tmp:
                    gff_dict[other].append(tmp[other])
                elif other==&#34;Strand&#34;:
                    gff_dict[other].append(ori)
    return gff_dict

def _read_and_reshape_bw(chrom, s, e, _bw, binsize):
    bw=pwg.open(_bw)
    val=bw.values(chrom, s, e)
    #print(val)
    val=np.array(val)
    #print(val.shape)
    val=val.reshape([-1,binsize]).mean(axis=1)
    #print(val.shape)
    _sum=np.sum(val)
    if _sum==0:
        mean=0
    else:
        mean=_sum/(e-s)
    # mean=bw.stats(chrom, s, e, exact=True)[0]
    # if mean==None:
    #     mean=0
    return val, mean 

def _read_bw_stats(chrom, s, e, _bw):
    bw=pwg.open(_bw)
    val=bw.stats(chrom, s, e)[0]
    if val==None:
        val=0
    return val 
def _read_and_reshape_np(s, e, bw, binsize):
    val=bw[s:e]
    #print(val)
    val=np.array(val)
    #print(val.shape)
    val=val.reshape([-1,binsize]).mean(axis=1)
    #print(val.shape)
    mean=val.mean()
    if mean==None:
        mean=0
    return val, mean

def _flatten_gen_comp(lst: List[Any]) -&gt; Iterable[Any]:
                &#34;&#34;&#34;Flatten a list using generators comprehensions.&#34;&#34;&#34;
                return (item
                        for sublist in lst
                        for item in sublist)
                
def _read_transcripts(_bw, _pos, binsize, extend):
    expected=2*extend//binsize
    bw=pwg.open(_bw)
    vals=[]
    for chrom, s, e in _pos:
        try:
            val=bw.values(chrom, s, e)
        except RuntimeError as e:
            print(chrom, s, e)
        vals.append(val)
        
    #print(val)
    val=np.nan_to_num(list(_flatten_gen_comp(vals)))
    
    
    
    val=val.reshape([-1,binsize]).mean(axis=1)
    if val.shape[0]&lt;expected:
        
        val =np.concatenate([val, np.zeros([expected-val.shape[0]])])
    return val
    
    
    
    
    
    </code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="omniplot.chipseq_utils.gff_parser"><code class="flex name class">
<span>class <span class="ident">gff_parser</span></span>
<span>(</span><span>gff)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class gff_parser():
    resolution=100000
    def __init__(self,gff):
        from operator import methodcaller
        time_start=time.time()
        home=os.path.expanduser(&#39;~&#39;)
        cache=os.path.join(home, &#34;omniplot_cache&#34;)
        if not os.path.isdir(cache):
            os.makedirs(cache)
        h, t=os.path.split(gff)
        gffchache=os.path.join(cache, t+&#34;.pickle&#34;)
        gffchache_info=os.path.join(cache, t+&#34;_info.pickle&#34;)
        if os.path.isfile(gffchache):
            with open(gffchache, &#34;rb&#34;) as fin:
                data=pickle.load(fin)
            with open(gffchache_info, &#34;rb&#34;) as fin:
                info=pickle.load(fin)
        else:
            data={}
            info={}
            with open(gff) as fin:
                for l in fin:
                    if l.startswith(&#34;#&#34;):
                        if l.startswith(&#34;##gff-version&#34;):
                            l=l.split()
                            info[&#34;gff-version&#34;]=l[1]
                        elif l.startswith(&#34;#description:&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;description&#34;]=l[1]
                        elif l.startswith(&#34;#provider:&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;provider&#34;]=l[1]
                        elif l.startswith(&#34;#format&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;format&#34;]=l[1]
                        elif l.startswith(&#34;#date&#34;):
                            l=l.strip(&#34;\n&#34;).split(&#34;: &#34;)
                            info[&#34;date&#34;]=l[1]
                        continue
                    chrom, source, kind, s, e, _, ori, _, meta=l.split()
                    s, e=int(s)-1, int(e)
                    meta=meta.split(&#34;;&#34;)
                    tmp={}
                    for m in meta:
                        k, v=m.split(&#34;=&#34;)
                        tmp[k]=v
                    

                    # if not chrom in data:
                    #     data[chrom]={&#34;pos&#34;:[], &#34;kind&#34;:[],&#34;ori&#34;:[],&#34;meta&#34;:[]}
                    #
                    #
                    # #tmp={k: v for k,v in map(methodcaller(&#34;split&#34;, &#34;=&#34;), meta)}
                    # data[chrom][&#34;pos&#34;].append([s, e])
                    # data[chrom][&#34;kind&#34;].append(kind)
                    # data[chrom][&#34;ori&#34;].append(ori)
                    # data[chrom][&#34;meta&#34;].append(tmp)        
                    if not chrom in data:
                        data[chrom]={}
                  
                    _s=self.resolution*(s//self.resolution)
                    if not _s in data[chrom]:
                        data[chrom][_s]={&#34;pos&#34;:[], &#34;kind&#34;:[],&#34;ori&#34;:[],&#34;meta&#34;:[]}
                    data[chrom][_s][&#34;pos&#34;].append([s, e])
                    data[chrom][_s][&#34;kind&#34;].append(kind)
                    data[chrom][_s][&#34;ori&#34;].append(ori)
                    data[chrom][_s][&#34;meta&#34;].append(tmp)  
            with open(gffchache, &#34;wb&#34;) as fin:
                pickle.dump(data, fin)
            with open(gffchache_info, &#34;wb&#34;) as fin:
                pickle.dump(info, fin)         
        self.data=data
        self.info=info
        print(time.time()-time_start)
    def get_genes(self, chrom: str,
                  start: int, 
                  end: int,
                  gene_type: set=set([&#34;protein_coding&#34;])) -&gt; list:
        _data=self.data[chrom]
        genes=[]
        
        
        _start=start//self.resolution
        _end=end//self.resolution
        for h in range(_start, _end+1):
            _h=self.resolution*h
            _data2=_data[_h]
            for i, (se, kind, meta, ori) in enumerate(zip(_data2[&#34;pos&#34;], _data2[&#34;kind&#34;], _data2[&#34;meta&#34;], _data2[&#34;ori&#34;])):
                s, e=se
                if e &lt; start:
                    continue
                if s&gt;end:
                    break
                if kind==&#34;gene&#34; and meta[&#34;gene_type&#34;] in gene_type:
                    
                
                    if start&lt;=s and e&lt;=end:
                        genename=meta[&#34;gene_name&#34;]
                        genes.append([genename,s,e,ori])
                    elif start&lt;=s&lt;=end and end&lt;e:
                        genename=meta[&#34;gene_name&#34;]
                        genes.append([genename,s,end,ori])
                    elif s&lt;start and start&lt;=e&lt;=end  and meta[&#34;gene_type&#34;] in gene_type:
                        genename=meta[&#34;gene_name&#34;]
                        genes.append([genename,start,e,ori])
        
        
        # _data=self.data[chrom]
        # genes=[]
        # for i, (se, kind, meta, ori) in enumerate(zip(_data[&#34;pos&#34;], _data[&#34;kind&#34;], _data[&#34;meta&#34;], _data[&#34;ori&#34;])):
        #     s, e=se
        #     if e &lt; start:
        #         continue
        #     if s&gt;end:
        #         break
        #     if kind==&#34;gene&#34; and meta[&#34;gene_type&#34;] in gene_type:
        #
        #
        #         if start&lt;=s and e&lt;=end:
        #             genename=meta[&#34;gene_name&#34;]
        #             genes.append([genename,s,e,ori])
        #         elif start&lt;=s&lt;=end and end&lt;e:
        #             genename=meta[&#34;gene_name&#34;]
        #             genes.append([genename,s,end,ori])
        #         elif s&lt;start and start&lt;=e&lt;=end  and meta[&#34;gene_type&#34;] in gene_type:
        #             genename=meta[&#34;gene_name&#34;]
        #             genes.append([genename,start,e,ori])
                
            

        return genes
    def get_tss(self, gene_type=set([&#34;protein_coding&#34;])):
        
        tss={}
        
        for chrom, _data in self.data.items():
            tss[chrom]={&#34;pos&#34;:[],&#34;genes&#34;:[]}
            for i, (s,e) in enumerate(_data[&#34;pos&#34;]):
                if _data[&#34;kind&#34;][i]==&#34;gene&#34; and _data[&#34;meta&#34;][i][&#34;gene_type&#34;] in gene_type:
                    tss[chrom][&#34;genes&#34;].append(_data[&#34;meta&#34;][i][&#34;gene_name&#34;])
                    if _data[&#34;ori&#34;][i]==&#34;+&#34;:
                        tss[chrom][&#34;pos&#34;].append(s)
                    else:
                        tss[chrom][&#34;pos&#34;].append(e)
        
        return tss 
    
    def get_tss_extend(self, gene_type=set([&#34;protein_coding&#34;]), extend=5000):
        
        tss={}
        
        for chrom, _data in self.data.items():
            tss[chrom]={&#34;pos&#34;:[],&#34;genes&#34;:[]}
            for i, (s,e) in enumerate(_data[&#34;pos&#34;]):
                if _data[&#34;kind&#34;][i]==&#34;gene&#34; and _data[&#34;meta&#34;][i][&#34;gene_type&#34;] in gene_type:
                    tss[chrom][&#34;genes&#34;].append(_data[&#34;meta&#34;][i][&#34;gene_name&#34;])
                    if _data[&#34;ori&#34;][i]==&#34;+&#34;:
                        tss[chrom][&#34;pos&#34;].append([s-extend,s+extend])
                    else:
                        tss[chrom][&#34;pos&#34;].append([e-extend,e+extend])
        
        return tss
    
    def closest_genes(self, peaks, K=1):
        tss=self.get_tss()
        
        if type(peaks)==list:
            _peaks={}
            for chrom, s, e in peaks:
                if not chrom in _peaks:
                    _peaks[chrom]={&#34;se&#34;:[],&#34;center&#34;:[]}
                _peaks[chrom][&#34;center&#34;].append((s+e)/2)
                _peaks[chrom][&#34;se&#34;].append([s,e])
            peaks=_peaks
        genes_peaks=[]
        i=0
        for chrom in tss.keys():
            if not chrom in peaks:
                continue
            _tss=np.array(tss[chrom][&#34;pos&#34;]).reshape([-1,1])
            _genes=np.array(tss[chrom][&#34;genes&#34;])
            _peaks=np.array(peaks[chrom][&#34;center&#34;]).reshape([-1,1])
            tree = BallTree(_tss)
            dist, ind = tree.query(_peaks, k=K) 
            for _ind, _dist, (s,e) in zip(list(ind), list(dist),peaks[chrom][&#34;se&#34;]):
                _closest=_genes[_ind]
                genes_peaks.append([chrom, s, e, &#34;,&#34;.join(list(_closest)), &#34;,&#34;.join(map(str, list(_dist)))])
        return genes_peaks</code></pre>
</details>
<h3>Class variables</h3>
<dl>
<dt id="omniplot.chipseq_utils.gff_parser.resolution"><code class="name">var <span class="ident">resolution</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="omniplot.chipseq_utils.gff_parser.closest_genes"><code class="name flex">
<span>def <span class="ident">closest_genes</span></span>(<span>self, peaks, K=1)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def closest_genes(self, peaks, K=1):
    tss=self.get_tss()
    
    if type(peaks)==list:
        _peaks={}
        for chrom, s, e in peaks:
            if not chrom in _peaks:
                _peaks[chrom]={&#34;se&#34;:[],&#34;center&#34;:[]}
            _peaks[chrom][&#34;center&#34;].append((s+e)/2)
            _peaks[chrom][&#34;se&#34;].append([s,e])
        peaks=_peaks
    genes_peaks=[]
    i=0
    for chrom in tss.keys():
        if not chrom in peaks:
            continue
        _tss=np.array(tss[chrom][&#34;pos&#34;]).reshape([-1,1])
        _genes=np.array(tss[chrom][&#34;genes&#34;])
        _peaks=np.array(peaks[chrom][&#34;center&#34;]).reshape([-1,1])
        tree = BallTree(_tss)
        dist, ind = tree.query(_peaks, k=K) 
        for _ind, _dist, (s,e) in zip(list(ind), list(dist),peaks[chrom][&#34;se&#34;]):
            _closest=_genes[_ind]
            genes_peaks.append([chrom, s, e, &#34;,&#34;.join(list(_closest)), &#34;,&#34;.join(map(str, list(_dist)))])
    return genes_peaks</code></pre>
</details>
</dd>
<dt id="omniplot.chipseq_utils.gff_parser.get_genes"><code class="name flex">
<span>def <span class="ident">get_genes</span></span>(<span>self, chrom: str, start: int, end: int, gene_type: set = {'protein_coding'}) ‑> list</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_genes(self, chrom: str,
              start: int, 
              end: int,
              gene_type: set=set([&#34;protein_coding&#34;])) -&gt; list:
    _data=self.data[chrom]
    genes=[]
    
    
    _start=start//self.resolution
    _end=end//self.resolution
    for h in range(_start, _end+1):
        _h=self.resolution*h
        _data2=_data[_h]
        for i, (se, kind, meta, ori) in enumerate(zip(_data2[&#34;pos&#34;], _data2[&#34;kind&#34;], _data2[&#34;meta&#34;], _data2[&#34;ori&#34;])):
            s, e=se
            if e &lt; start:
                continue
            if s&gt;end:
                break
            if kind==&#34;gene&#34; and meta[&#34;gene_type&#34;] in gene_type:
                
            
                if start&lt;=s and e&lt;=end:
                    genename=meta[&#34;gene_name&#34;]
                    genes.append([genename,s,e,ori])
                elif start&lt;=s&lt;=end and end&lt;e:
                    genename=meta[&#34;gene_name&#34;]
                    genes.append([genename,s,end,ori])
                elif s&lt;start and start&lt;=e&lt;=end  and meta[&#34;gene_type&#34;] in gene_type:
                    genename=meta[&#34;gene_name&#34;]
                    genes.append([genename,start,e,ori])
    
    
    # _data=self.data[chrom]
    # genes=[]
    # for i, (se, kind, meta, ori) in enumerate(zip(_data[&#34;pos&#34;], _data[&#34;kind&#34;], _data[&#34;meta&#34;], _data[&#34;ori&#34;])):
    #     s, e=se
    #     if e &lt; start:
    #         continue
    #     if s&gt;end:
    #         break
    #     if kind==&#34;gene&#34; and meta[&#34;gene_type&#34;] in gene_type:
    #
    #
    #         if start&lt;=s and e&lt;=end:
    #             genename=meta[&#34;gene_name&#34;]
    #             genes.append([genename,s,e,ori])
    #         elif start&lt;=s&lt;=end and end&lt;e:
    #             genename=meta[&#34;gene_name&#34;]
    #             genes.append([genename,s,end,ori])
    #         elif s&lt;start and start&lt;=e&lt;=end  and meta[&#34;gene_type&#34;] in gene_type:
    #             genename=meta[&#34;gene_name&#34;]
    #             genes.append([genename,start,e,ori])
            
        

    return genes</code></pre>
</details>
</dd>
<dt id="omniplot.chipseq_utils.gff_parser.get_tss"><code class="name flex">
<span>def <span class="ident">get_tss</span></span>(<span>self, gene_type={'protein_coding'})</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_tss(self, gene_type=set([&#34;protein_coding&#34;])):
    
    tss={}
    
    for chrom, _data in self.data.items():
        tss[chrom]={&#34;pos&#34;:[],&#34;genes&#34;:[]}
        for i, (s,e) in enumerate(_data[&#34;pos&#34;]):
            if _data[&#34;kind&#34;][i]==&#34;gene&#34; and _data[&#34;meta&#34;][i][&#34;gene_type&#34;] in gene_type:
                tss[chrom][&#34;genes&#34;].append(_data[&#34;meta&#34;][i][&#34;gene_name&#34;])
                if _data[&#34;ori&#34;][i]==&#34;+&#34;:
                    tss[chrom][&#34;pos&#34;].append(s)
                else:
                    tss[chrom][&#34;pos&#34;].append(e)
    
    return tss </code></pre>
</details>
</dd>
<dt id="omniplot.chipseq_utils.gff_parser.get_tss_extend"><code class="name flex">
<span>def <span class="ident">get_tss_extend</span></span>(<span>self, gene_type={'protein_coding'}, extend=5000)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_tss_extend(self, gene_type=set([&#34;protein_coding&#34;]), extend=5000):
    
    tss={}
    
    for chrom, _data in self.data.items():
        tss[chrom]={&#34;pos&#34;:[],&#34;genes&#34;:[]}
        for i, (s,e) in enumerate(_data[&#34;pos&#34;]):
            if _data[&#34;kind&#34;][i]==&#34;gene&#34; and _data[&#34;meta&#34;][i][&#34;gene_type&#34;] in gene_type:
                tss[chrom][&#34;genes&#34;].append(_data[&#34;meta&#34;][i][&#34;gene_name&#34;])
                if _data[&#34;ori&#34;][i]==&#34;+&#34;:
                    tss[chrom][&#34;pos&#34;].append([s-extend,s+extend])
                else:
                    tss[chrom][&#34;pos&#34;].append([e-extend,e+extend])
    
    return tss</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="omniplot" href="index.html">omniplot</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="omniplot.chipseq_utils.gff_parser" href="#omniplot.chipseq_utils.gff_parser">gff_parser</a></code></h4>
<ul class="">
<li><code><a title="omniplot.chipseq_utils.gff_parser.closest_genes" href="#omniplot.chipseq_utils.gff_parser.closest_genes">closest_genes</a></code></li>
<li><code><a title="omniplot.chipseq_utils.gff_parser.get_genes" href="#omniplot.chipseq_utils.gff_parser.get_genes">get_genes</a></code></li>
<li><code><a title="omniplot.chipseq_utils.gff_parser.get_tss" href="#omniplot.chipseq_utils.gff_parser.get_tss">get_tss</a></code></li>
<li><code><a title="omniplot.chipseq_utils.gff_parser.get_tss_extend" href="#omniplot.chipseq_utils.gff_parser.get_tss_extend">get_tss_extend</a></code></li>
<li><code><a title="omniplot.chipseq_utils.gff_parser.resolution" href="#omniplot.chipseq_utils.gff_parser.resolution">resolution</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>